{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eb5d9dfd",
   "metadata": {},
   "source": [
    "# Overview\n",
    "This file contains an example workflow to create a customized TIPP reference package, using some example genes included under `example_data/` with their necessary files:\n",
    "1. the alignment,\n",
    "2. the taxid of sequences,\n",
    "3. the mapping between each sequence and its taxid.\n",
    "\n",
    "----------------------\n",
    "# Required Packages\n",
    "```\n",
    "taxtastic>=0.10.0\n",
    "dendropy>=4.5.2\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b47d2ea0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import os, sys\n",
    "from subprocess import Popen, PIPE, STDOUT\n",
    "\n",
    "# number of threads to use for raxml-ng and fasttree-2\n",
    "num_threads = 4\n",
    "\n",
    "# initialize work directory\n",
    "workdir = 'example_data'\n",
    "\n",
    "# where to write the TIPP reference package to\n",
    "refpkg_dir = 'custom_tipp_refpkg'\n",
    "\n",
    "# initialize the list of genes to use, which are assumed to be under the working directory\n",
    "genes = ['RplA_COG0081', 'RplB_COG0090', 'RplD_COG0088', 'RpsC_COG0092', 'RpsM_COG0099']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bc6b2d8",
   "metadata": {},
   "source": [
    "# Needed files\n",
    "In this example, we are only using one gene to create our customized TIPP reference package, included under `example_data/RplO_COG0200`. Remember that there are 4 files that we need to prepare (one for global use, and three for each gene):\n",
    "1. `ncbi_taxonomy.db  - (global) NCBI taxonomy file created with taxtastic`\n",
    "2. `est.aln.nuc.fasta[.gz] - (each gene) alignment file (can be gzipped)`\n",
    "3. `species.txt       - (each gene) taxids of sequences in alignment, in the same order as in the alignment`\n",
    "4. `species.mapping   - (each gene) mapping between sequences and taxids, in the same order as in the alignment`\n",
    "\n",
    "The three files for the example gene are included, and the NCBI taxonomy database can be built below."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a65e8e05",
   "metadata": {},
   "source": [
    "# Step 0: NCBI taxonomy database - `ncbi_taxonomy.db`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1b0fd8fa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# download the latest NCBI taxdmp.zip\n",
    "url = \"https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdmp.zip\"\n",
    "\n",
    "# run taxtastic to create a database locally\n",
    "if not os.path.exists(f\"{workdir}/ncbi_taxonomy.db\"):\n",
    "    cmd = f\"wget -P {workdir} {url}\"\n",
    "    os.system(cmd)\n",
    "\n",
    "    cmd = f\"taxit new_database -z {workdir}/taxdmp.zip\"\n",
    "    os.system(cmd)\n",
    "\n",
    "    # move the created ncbi_taxonomy.db file to workdir\n",
    "    os.system(f\"mv ncbi_taxonomy.db {workdir}/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cb805e7",
   "metadata": {},
   "source": [
    "# Step 1: Update species taxids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "104ea3ec",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "db_path = os.path.join(workdir, 'ncbi_taxonomy.db')\n",
    "\n",
    "# two files to update\n",
    "old2new_map = {'species.txt': 'species.updated.txt',\n",
    "               'species.mapping': 'species.updated.mapping',\n",
    "              }\n",
    "\n",
    "for gene in genes:\n",
    "    indir = os.path.join(workdir, gene)\n",
    "\n",
    "    for infile, outfile in old2new_map.items():\n",
    "        inpath = os.path.join(indir, infile)\n",
    "        outpath = os.path.join(indir, outfile)\n",
    "        # run taxit update_taxids\n",
    "        cmd = f\"taxit update_taxids {inpath} {db_path} -o {outpath}\"\n",
    "        os.system(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76e31710",
   "metadata": {},
   "source": [
    "# Step 2: Get taxonomy table from the included species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b05f435d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "db_path = os.path.join(workdir, 'ncbi_taxonomy.db')\n",
    "\n",
    "for gene in genes:\n",
    "    indir = os.path.join(workdir, gene)\n",
    "    \n",
    "    # get taxonomy.table with taxit\n",
    "    species_path = os.path.join(indir, 'species.updated.txt')\n",
    "    taxtable_path = os.path.join(indir, 'taxonomy.table')\n",
    "    cmd = f\"taxit taxtable {db_path} -i {species_path} -o {taxtable_path}\"\n",
    "    os.system(cmd)\n",
    "    \n",
    "    # get a cleaned version of the taxonomy table by removing redundant columns\n",
    "    alltaxon_path = os.path.join(indir, 'all_taxon.taxonomy')\n",
    "    cmd = f\"python clean_taxonomy_table.py {taxtable_path} {alltaxon_path}\"\n",
    "    os.system(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f976fc8f",
   "metadata": {},
   "source": [
    "# Step 3: Initialize taxonomy tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7eb43c4f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>>> RplA_COG0081\n",
      "Child taxids at root (131567): 2,2157\n",
      ">>>> RplB_COG0090\n",
      "Child taxids at root (131567): 2\n",
      "WARNING: node (taxid 1275) unifurcates and have assigned sequences!\n",
      "\tRe-assigning sequences with taxid 136273 to 1275\n",
      "Reassigning taxids to sequences, backing up example_data/RplB_COG0090/species.updated.txt and example_data/RplB_COG0090/species.updated.mapping ...\n",
      "GCF_001483755.1_NZ_LQBK01000002.1_cds_WP_058872530.1_238 taxid reassigned from 136273 to 1275\n",
      "Reassigned taxids for 1/5510 sequences\n",
      ">>>> RplD_COG0088\n",
      "Child taxids at root (131567): 2\n",
      "WARNING: node (taxid 561879) unifurcates and have assigned sequences!\n",
      "\tRe-assigning sequences with taxid 1178541 to 561879\n",
      "WARNING: node (taxid 1421) unifurcates and have assigned sequences!\n",
      "\tRe-assigning sequences with taxid 444177 to 1421\n",
      "Reassigning taxids to sequences, backing up example_data/RplD_COG0088/species.updated.txt and example_data/RplD_COG0088/species.updated.mapping ...\n",
      "GCF_000017965.1_NC_010382.1_cds_WP_012296062.1_4284 taxid reassigned from 444177 to 1421\n",
      "GCF_000691165.1_NZ_ASJD01000012.1_cds_WP_024425598.1_3451 taxid reassigned from 1178541 to 561879\n",
      "Reassigned taxids for 2/5462 sequences\n",
      ">>>> RpsC_COG0092\n",
      "Child taxids at root (131567): 2,2157\n",
      ">>>> RpsM_COG0099\n",
      "Child taxids at root (131567): 2\n"
     ]
    }
   ],
   "source": [
    "for gene in genes:\n",
    "    print('>>>>', gene)\n",
    "    indir = os.path.join(workdir, gene)\n",
    "\n",
    "    # run script build_taxonomic_tree.py\n",
    "    # NOTE: both species.updated.txt and species.updated.mapping may be updated\n",
    "    #       if there are unifurication detected in the taxonomy tree.\n",
    "    #       e.g., if we have the following taxids: 57493 --> 1275 --> 136273\n",
    "    #       then, both nodes 1275 and 136273 have sequences assigne to them.\n",
    "    #       However, since node 1275 only has one child, it will be \"contracted\" when\n",
    "    #       eliminating unifurication. Thus, sequences assigned to it may not appear\n",
    "    #       in the unrefined.taxonomy.renamed file and leads to error when running RAxML-ng\n",
    "    #\n",
    "    #       This is a rare event, and the current solution is to modify nodes so that if we find\n",
    "    #       a unifuricating node with taxid and sequences are assigned to it, we will contract all\n",
    "    #       branches below this node and re-assign the sequences belonging to leaves of this subtree\n",
    "    #       (rooted at the unifuricating node) to this node.\n",
    "    #       Then, both species.updated.txt and species.updated.mapping would be updated to reflect\n",
    "    #       the reassignment of taxids (to the sequences that are affected).\n",
    "    species_path = os.path.join(indir, 'species.updated.txt')\n",
    "    mapping_path = os.path.join(indir, 'species.updated.mapping')\n",
    "    alltaxon_path = os.path.join(indir, 'all_taxon.taxonomy')\n",
    "    unrefined_path = os.path.join(indir, 'unrefined.taxonomy')\n",
    "    cmd = f\"python build_taxonomic_tree.py {alltaxon_path} {species_path} {mapping_path} {unrefined_path}\"\n",
    "    p = Popen(cmd, stdout=PIPE, stderr=STDOUT, text=True, shell=True)\n",
    "    for line in p.stdout:\n",
    "        print(line.replace('\\n', ''))\n",
    "    \n",
    "    # rename the taxonomy tree with sequence names from this gene\n",
    "    # with script build_unrefined_tree.pl\n",
    "    taxonomy_path = os.path.join(indir, 'unrefined.taxonomy.renamed')\n",
    "    cmd = f\"perl build_unrefined_tree.pl {mapping_path} {unrefined_path} {taxonomy_path}\"\n",
    "    os.system(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f44bd96b",
   "metadata": {},
   "source": [
    "# (for tipp3) Get RAxML-ng maximum likelihood tree and model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "53ee7136",
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in genes:\n",
    "    indir = os.path.join(workdir, gene)\n",
    "    \n",
    "    # masked the alignment first with >= 95% gaps\n",
    "    aln_path = os.path.join(indir, 'est.aln.nuc.fasta.gz') \n",
    "    masked_path = os.path.join(indir, 'est.aln.nuc.masked95.fasta')\n",
    "    thres = 0.95\n",
    "    cmd = f\"python mask_alignment.py {aln_path} {masked_path} {thres}\"\n",
    "    os.system(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6b3a7451",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for gene in genes:\n",
    "    # run raxml-ng\n",
    "    masked_path = os.path.join(indir, 'est.aln.nuc.masked95.fasta')\n",
    "    check_path = os.path.join(indir, 'est.raxml.bestTree')\n",
    "    if not os.path.exists(check_path):\n",
    "        # resolve polytomies in the taxonomy tree and remove internal node labels to run RAxML-ng with\n",
    "        taxonomy_path = os.path.join(indir, 'unrefined.taxonomy.renamed')\n",
    "        starting_tree_path = os.path.join(indir, 'raxml.starting.tre')\n",
    "        cmd = f\"python resolve_polytomies.py {taxonomy_path} {starting_tree_path}\"\n",
    "        os.system(cmd)\n",
    "\n",
    "        # Run RAxML-ng with the starting tree and the original taxonomy tree as the constraint\n",
    "        cmd = ['raxml-ng', '--msa', masked_path, '--model GTR+G',\n",
    "               '--tree', starting_tree_path, '--tree-constraint', taxonomy_path,\n",
    "               '--brlen scaled',\n",
    "               '--force perf_threads', '--redo',\n",
    "               f'--prefix {indir}/est',\n",
    "               '--threads', str(num_threads)]\n",
    "\n",
    "        p = Popen(' '.join(cmd), stdout=PIPE, stderr=STDOUT, text=True, shell=True)\n",
    "        for line in p.stdout:\n",
    "            if not line.startswith('WARNING'):\n",
    "                print(line.replace('\\n', ''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a6096dee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>>> RplA_COG0081\n",
      "reseeding example_data/RplA_COG0081/est.raxml.bestTree\n",
      "relabeling example_data/RplA_COG0081/est.raxml.bestTree\n",
      "\tmapped 637\n",
      "\tmissing 0\n",
      "sanity check correctness True\n",
      "\n",
      ">>>> RplB_COG0090\n",
      "reseeding example_data/RplB_COG0090/est.raxml.bestTree (only one superkingdom group)\n",
      "relabeling example_data/RplB_COG0090/est.raxml.bestTree\n",
      "\tmapped 653\n",
      "\tmissing 0\n",
      "sanity check correctness True\n",
      "\n",
      ">>>> RplD_COG0088\n",
      "reseeding example_data/RplD_COG0088/est.raxml.bestTree (only one superkingdom group)\n",
      "relabeling example_data/RplD_COG0088/est.raxml.bestTree\n",
      "\tmapped 635\n",
      "\tmissing 0\n",
      "sanity check correctness True\n",
      "\n",
      ">>>> RpsC_COG0092\n",
      "reseeding example_data/RpsC_COG0092/est.raxml.bestTree\n",
      "relabeling example_data/RpsC_COG0092/est.raxml.bestTree\n",
      "\tmapped 641\n",
      "\tmissing 0\n",
      "sanity check correctness True\n",
      "\n",
      ">>>> RpsM_COG0099\n",
      "reseeding example_data/RpsM_COG0099/est.raxml.bestTree (only one superkingdom group)\n",
      "relabeling example_data/RpsM_COG0099/est.raxml.bestTree\n",
      "\tmapped 664\n",
      "\tmissing 0\n",
      "sanity check correctness True\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# add back the taxonomy information to the raxml output tree, and reroot the tree at taxid 131567\n",
    "for gene in genes:\n",
    "    print('>>>>', gene)\n",
    "    indir = os.path.join(workdir, gene)\n",
    "    \n",
    "    # reroot the raxml-ng bestTree to 131567 (cellular organism) and relabel the tree\n",
    "    root = 131567\n",
    "    best_tree_path = os.path.join(indir, 'est.raxml.bestTree')\n",
    "    taxonomy_path = os.path.join(indir, 'unrefined.taxonomy.renamed')\n",
    "    rooted_tree_path = os.path.join(indir, 'est.raxml.bestTree.rooted')\n",
    "    cmd = f\"python relabel_tree.py {taxonomy_path} {best_tree_path} {rooted_tree_path} {root}\"\n",
    "    ret = os.popen(cmd).read()\n",
    "    print(ret)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f69a0018",
   "metadata": {},
   "source": [
    "# (for tipp3-accurate) Get FastTree-2 maximum likelihood tree and log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ff0e98bd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>>> RplA_COG0081\n",
      "FastTree Version 2.1.11 SSE3, OpenMP (12 threads)\n",
      "Alignment: standard input\n",
      "Nucleotide distances: Jukes-Cantor Joins: balanced Support: none\n",
      "Start at tree from example_data/RplA_COG0081/est.raxml.bestTree.rooted (no NNI) (no SPR)\n",
      "ML Model: Generalized Time-Reversible, CAT approximation with 20 rate categories\n",
      "Total branch-length 174.238 after 0.13 sec\n",
      "      0.14 seconds: ML Lengths 1 of 1897 splits\n",
      "      0.26 seconds: ML Lengths 401 of 1897 splits\n",
      "      0.38 seconds: ML Lengths 801 of 1897 splits\n",
      "      0.51 seconds: ML Lengths 1201 of 1897 splits\n",
      "      0.63 seconds: ML Lengths 1601 of 1897 splits\n",
      "1 rounds ML lengths: LogLk = -540809.088 Max-change 0.2599 Time 0.74\n",
      "      0.74 seconds: Optimizing GTR model, step 1 of 12\n",
      "      1.28 seconds: Optimizing GTR model, step 2 of 12\n",
      "      1.66 seconds: Optimizing GTR model, step 3 of 12\n",
      "      2.03 seconds: Optimizing GTR model, step 4 of 12\n",
      "      2.49 seconds: Optimizing GTR model, step 5 of 12\n",
      "      3.07 seconds: Optimizing GTR model, step 6 of 12\n",
      "      3.70 seconds: Optimizing GTR model, step 7 of 12\n",
      "      4.07 seconds: Optimizing GTR model, step 8 of 12\n",
      "      4.40 seconds: Optimizing GTR model, step 9 of 12\n",
      "      4.82 seconds: Optimizing GTR model, step 10 of 12\n",
      "      5.19 seconds: Optimizing GTR model, step 11 of 12\n",
      "      5.61 seconds: Optimizing GTR model, step 12 of 12\n",
      "GTR Frequencies: 0.2634 0.2490 0.2750 0.2125\n",
      "GTR rates(ac ag at cg ct gt) 1.3110 2.2821 2.4271 1.9328 4.4521 1.0000\n",
      "      6.02 seconds: ML Lengths 1 of 1897 splits\n",
      "      6.14 seconds: ML Lengths 301 of 1897 splits\n",
      "      6.27 seconds: ML Lengths 601 of 1897 splits\n",
      "      6.39 seconds: ML Lengths 901 of 1897 splits\n",
      "      6.52 seconds: ML Lengths 1201 of 1897 splits\n",
      "      6.65 seconds: ML Lengths 1501 of 1897 splits\n",
      "      6.77 seconds: ML Lengths 1801 of 1897 splits\n",
      "      6.90 seconds: Site likelihoods with rate category 2 of 20\n",
      "      7.03 seconds: Site likelihoods with rate category 5 of 20\n",
      "      7.16 seconds: Site likelihoods with rate category 8 of 20\n",
      "      7.29 seconds: Site likelihoods with rate category 11 of 20\n",
      "      7.42 seconds: Site likelihoods with rate category 14 of 20\n",
      "      7.55 seconds: Site likelihoods with rate category 17 of 20\n",
      "      7.68 seconds: Site likelihoods with rate category 20 of 20\n",
      "Switched to using 20 rate categories (CAT approximation)\n",
      "Rate categories were divided by 1.029 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "      7.79 seconds: ML Lengths 101 of 1897 splits\n",
      "      7.92 seconds: ML Lengths 401 of 1897 splits\n",
      "      8.05 seconds: ML Lengths 701 of 1897 splits\n",
      "      8.19 seconds: ML Lengths 1001 of 1897 splits\n",
      "      8.32 seconds: ML Lengths 1301 of 1897 splits\n",
      "      8.46 seconds: ML Lengths 1601 of 1897 splits\n",
      "2 rounds ML lengths: LogLk = -470661.191 Max-change 1.6922 Time 8.61\n",
      "      8.60 seconds: ML Lengths 1 of 1897 splits\n",
      "      8.73 seconds: ML Lengths 301 of 1897 splits\n",
      "      8.85 seconds: ML Lengths 601 of 1897 splits\n",
      "      8.98 seconds: ML Lengths 901 of 1897 splits\n",
      "      9.11 seconds: ML Lengths 1201 of 1897 splits\n",
      "      9.24 seconds: ML Lengths 1501 of 1897 splits\n",
      "      9.37 seconds: ML Lengths 1801 of 1897 splits\n",
      "3 rounds ML lengths: LogLk = -470643.338 Max-change 0.0942 Time 9.43\n",
      "      9.51 seconds: ML Lengths 201 of 1897 splits\n",
      "      9.63 seconds: ML Lengths 501 of 1897 splits\n",
      "      9.76 seconds: ML Lengths 801 of 1897 splits\n",
      "      9.89 seconds: ML Lengths 1101 of 1897 splits\n",
      "     10.01 seconds: ML Lengths 1401 of 1897 splits\n",
      "     10.14 seconds: ML Lengths 1701 of 1897 splits\n",
      "4 rounds ML lengths: LogLk = -470641.892 Max-change 0.0176 Time 10.24\n",
      "     10.28 seconds: ML Lengths 101 of 1897 splits\n",
      "     10.40 seconds: ML Lengths 401 of 1897 splits\n",
      "     10.52 seconds: ML Lengths 701 of 1897 splits\n",
      "     10.65 seconds: ML Lengths 1001 of 1897 splits\n",
      "     10.78 seconds: ML Lengths 1301 of 1897 splits\n",
      "     10.90 seconds: ML Lengths 1601 of 1897 splits\n",
      "5 rounds ML lengths: LogLk = -470641.722 Max-change 0.0048 Time 11.04\n",
      "     11.04 seconds: ML Lengths 1 of 1897 splits\n",
      "     11.16 seconds: ML Lengths 301 of 1897 splits\n",
      "     11.29 seconds: ML Lengths 601 of 1897 splits\n",
      "     11.41 seconds: ML Lengths 901 of 1897 splits\n",
      "     11.54 seconds: ML Lengths 1201 of 1897 splits\n",
      "     11.67 seconds: ML Lengths 1501 of 1897 splits\n",
      "     11.79 seconds: ML Lengths 1801 of 1897 splits\n",
      "6 rounds ML lengths: LogLk = -470641.688 Max-change 0.0017 Time 11.85\n",
      "     11.93 seconds: ML Lengths 201 of 1897 splits\n",
      "     12.05 seconds: ML Lengths 501 of 1897 splits\n",
      "     12.18 seconds: ML Lengths 801 of 1897 splits\n",
      "     12.30 seconds: ML Lengths 1101 of 1897 splits\n",
      "     12.43 seconds: ML Lengths 1401 of 1897 splits\n",
      "     12.55 seconds: ML Lengths 1701 of 1897 splits\n",
      "7 rounds ML lengths: LogLk = -470641.678 Max-change 0.0010 (converged) Time 12.65\n",
      "     12.69 seconds: Site likelihoods with rate category 1 of 20\n",
      "     12.83 seconds: Site likelihoods with rate category 4 of 20\n",
      "     12.96 seconds: Site likelihoods with rate category 7 of 20\n",
      "     13.10 seconds: Site likelihoods with rate category 10 of 20\n",
      "     13.23 seconds: Site likelihoods with rate category 13 of 20\n",
      "     13.36 seconds: Site likelihoods with rate category 16 of 20\n",
      "     13.50 seconds: Site likelihoods with rate category 19 of 20\n",
      "Gamma(20) LogLk = -472351.494 alpha = 0.633 rescaling lengths by 1.212\n",
      "Total time: 13.59 seconds Unique: 1899/2749 Bad splits: 0/0\n",
      ">>>> RplB_COG0090\n",
      "FastTree Version 2.1.11 SSE3, OpenMP (12 threads)\n",
      "Alignment: standard input\n",
      "Nucleotide distances: Jukes-Cantor Joins: balanced Support: none\n",
      "Start at tree from example_data/RplB_COG0090/est.raxml.bestTree.rooted (no NNI) (no SPR)\n",
      "ML Model: Generalized Time-Reversible, CAT approximation with 20 rate categories\n",
      "Ignored unknown character X (seen 1 times)\n",
      "Total branch-length 162.417 after 0.16 sec\n",
      "      0.17 seconds: ML Lengths 1 of 1878 splits\n",
      "      0.28 seconds: ML Lengths 301 of 1878 splits\n",
      "      0.38 seconds: ML Lengths 601 of 1878 splits\n",
      "      0.51 seconds: ML Lengths 901 of 1878 splits\n",
      "      0.62 seconds: ML Lengths 1201 of 1878 splits\n",
      "      0.74 seconds: ML Lengths 1501 of 1878 splits\n",
      "      0.85 seconds: ML Lengths 1801 of 1878 splits\n",
      "1 rounds ML lengths: LogLk = -611123.225 Max-change 0.2415 Time 0.90\n",
      "      1.54 seconds: Optimizing GTR model, step 2 of 12\n",
      "      2.03 seconds: Optimizing GTR model, step 3 of 12\n",
      "      2.53 seconds: Optimizing GTR model, step 4 of 12\n",
      "      3.21 seconds: Optimizing GTR model, step 5 of 12\n",
      "      3.90 seconds: Optimizing GTR model, step 6 of 12\n",
      "      4.64 seconds: Optimizing GTR model, step 7 of 12\n",
      "      5.08 seconds: Optimizing GTR model, step 8 of 12\n",
      "      5.47 seconds: Optimizing GTR model, step 9 of 12\n",
      "      5.97 seconds: Optimizing GTR model, step 10 of 12\n",
      "      6.41 seconds: Optimizing GTR model, step 11 of 12\n",
      "      6.90 seconds: Optimizing GTR model, step 12 of 12\n",
      "GTR Frequencies: 0.2541 0.2654 0.2753 0.2053\n",
      "GTR rates(ac ag at cg ct gt) 1.2224 2.1273 2.5697 1.5889 4.7629 1.0000\n",
      "      7.38 seconds: ML Lengths 1 of 1878 splits\n",
      "      7.48 seconds: ML Lengths 201 of 1878 splits\n",
      "      7.63 seconds: ML Lengths 501 of 1878 splits\n",
      "      7.73 seconds: ML Lengths 701 of 1878 splits\n",
      "      7.84 seconds: ML Lengths 901 of 1878 splits\n",
      "      7.94 seconds: ML Lengths 1101 of 1878 splits\n",
      "      8.04 seconds: ML Lengths 1301 of 1878 splits\n",
      "      8.15 seconds: ML Lengths 1501 of 1878 splits\n",
      "      8.25 seconds: ML Lengths 1701 of 1878 splits\n",
      "      8.39 seconds: Site likelihoods with rate category 1 of 20\n",
      "      8.49 seconds: Site likelihoods with rate category 3 of 20\n",
      "      8.59 seconds: Site likelihoods with rate category 5 of 20\n",
      "      8.70 seconds: Site likelihoods with rate category 7 of 20\n",
      "      8.80 seconds: Site likelihoods with rate category 9 of 20\n",
      "      8.90 seconds: Site likelihoods with rate category 11 of 20\n",
      "      9.00 seconds: Site likelihoods with rate category 13 of 20\n",
      "      9.10 seconds: Site likelihoods with rate category 15 of 20\n",
      "      9.20 seconds: Site likelihoods with rate category 17 of 20\n",
      "      9.30 seconds: Site likelihoods with rate category 19 of 20\n",
      "Switched to using 20 rate categories (CAT approximation)\n",
      "Rate categories were divided by 1.002 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "      9.43 seconds: ML Lengths 1 of 1878 splits\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      9.54 seconds: ML Lengths 201 of 1878 splits\n",
      "      9.65 seconds: ML Lengths 401 of 1878 splits\n",
      "      9.75 seconds: ML Lengths 601 of 1878 splits\n",
      "      9.86 seconds: ML Lengths 801 of 1878 splits\n",
      "      9.97 seconds: ML Lengths 1001 of 1878 splits\n",
      "     10.08 seconds: ML Lengths 1201 of 1878 splits\n",
      "     10.19 seconds: ML Lengths 1401 of 1878 splits\n",
      "     10.30 seconds: ML Lengths 1601 of 1878 splits\n",
      "     10.41 seconds: ML Lengths 1801 of 1878 splits\n",
      "2 rounds ML lengths: LogLk = -530784.133 Max-change 0.4836 Time 10.47\n",
      "     10.51 seconds: ML Lengths 101 of 1878 splits\n",
      "     10.62 seconds: ML Lengths 301 of 1878 splits\n",
      "     10.77 seconds: ML Lengths 601 of 1878 splits\n",
      "     10.87 seconds: ML Lengths 801 of 1878 splits\n",
      "     10.98 seconds: ML Lengths 1001 of 1878 splits\n",
      "     11.08 seconds: ML Lengths 1201 of 1878 splits\n",
      "     11.18 seconds: ML Lengths 1401 of 1878 splits\n",
      "     11.29 seconds: ML Lengths 1601 of 1878 splits\n",
      "     11.39 seconds: ML Lengths 1801 of 1878 splits\n",
      "3 rounds ML lengths: LogLk = -530772.382 Max-change 0.1235 Time 11.45\n",
      "     11.50 seconds: ML Lengths 101 of 1878 splits\n",
      "     11.65 seconds: ML Lengths 401 of 1878 splits\n",
      "     11.80 seconds: ML Lengths 701 of 1878 splits\n",
      "     11.90 seconds: ML Lengths 901 of 1878 splits\n",
      "     12.00 seconds: ML Lengths 1101 of 1878 splits\n",
      "     12.11 seconds: ML Lengths 1301 of 1878 splits\n",
      "     12.21 seconds: ML Lengths 1501 of 1878 splits\n",
      "     12.31 seconds: ML Lengths 1701 of 1878 splits\n",
      "4 rounds ML lengths: LogLk = -530771.736 Max-change 0.0343 Time 12.42\n",
      "     12.41 seconds: ML Lengths 1 of 1878 splits\n",
      "     12.52 seconds: ML Lengths 201 of 1878 splits\n",
      "     12.66 seconds: ML Lengths 501 of 1878 splits\n",
      "     12.76 seconds: ML Lengths 701 of 1878 splits\n",
      "     12.87 seconds: ML Lengths 901 of 1878 splits\n",
      "     12.97 seconds: ML Lengths 1101 of 1878 splits\n",
      "     13.07 seconds: ML Lengths 1301 of 1878 splits\n",
      "     13.17 seconds: ML Lengths 1501 of 1878 splits\n",
      "     13.28 seconds: ML Lengths 1701 of 1878 splits\n",
      "5 rounds ML lengths: LogLk = -530771.634 Max-change 0.0104 Time 13.38\n",
      "     13.38 seconds: ML Lengths 1 of 1878 splits\n",
      "     13.48 seconds: ML Lengths 201 of 1878 splits\n",
      "     13.63 seconds: ML Lengths 501 of 1878 splits\n",
      "     13.73 seconds: ML Lengths 701 of 1878 splits\n",
      "     13.83 seconds: ML Lengths 901 of 1878 splits\n",
      "     13.93 seconds: ML Lengths 1101 of 1878 splits\n",
      "     14.03 seconds: ML Lengths 1301 of 1878 splits\n",
      "     14.14 seconds: ML Lengths 1501 of 1878 splits\n",
      "     14.24 seconds: ML Lengths 1701 of 1878 splits\n",
      "6 rounds ML lengths: LogLk = -530771.612 Max-change 0.0028 Time 14.34\n",
      "     14.34 seconds: ML Lengths 1 of 1878 splits\n",
      "     14.44 seconds: ML Lengths 201 of 1878 splits\n",
      "     14.59 seconds: ML Lengths 501 of 1878 splits\n",
      "     14.69 seconds: ML Lengths 701 of 1878 splits\n",
      "     14.79 seconds: ML Lengths 901 of 1878 splits\n",
      "     14.90 seconds: ML Lengths 1101 of 1878 splits\n",
      "     15.00 seconds: ML Lengths 1301 of 1878 splits\n",
      "     15.10 seconds: ML Lengths 1501 of 1878 splits\n",
      "     15.20 seconds: ML Lengths 1701 of 1878 splits\n",
      "7 rounds ML lengths: LogLk = -530771.606 Max-change 0.0010 Time 15.31\n",
      "     15.30 seconds: ML Lengths 1 of 1878 splits\n",
      "     15.40 seconds: ML Lengths 201 of 1878 splits\n",
      "     15.55 seconds: ML Lengths 501 of 1878 splits\n",
      "     15.65 seconds: ML Lengths 701 of 1878 splits\n",
      "     15.76 seconds: ML Lengths 901 of 1878 splits\n",
      "     15.86 seconds: ML Lengths 1101 of 1878 splits\n",
      "     15.96 seconds: ML Lengths 1301 of 1878 splits\n",
      "     16.06 seconds: ML Lengths 1501 of 1878 splits\n",
      "     16.16 seconds: ML Lengths 1701 of 1878 splits\n",
      "8 rounds ML lengths: LogLk = -530771.604 Max-change 0.0005 (converged) Time 16.27\n",
      "     16.32 seconds: Site likelihoods with rate category 1 of 20\n",
      "     16.42 seconds: Site likelihoods with rate category 3 of 20\n",
      "     16.53 seconds: Site likelihoods with rate category 5 of 20\n",
      "     16.63 seconds: Site likelihoods with rate category 7 of 20\n",
      "     16.74 seconds: Site likelihoods with rate category 9 of 20\n",
      "     16.84 seconds: Site likelihoods with rate category 11 of 20\n",
      "     16.95 seconds: Site likelihoods with rate category 13 of 20\n",
      "     17.05 seconds: Site likelihoods with rate category 15 of 20\n",
      "     17.16 seconds: Site likelihoods with rate category 17 of 20\n",
      "     17.27 seconds: Site likelihoods with rate category 19 of 20\n",
      "Gamma(20) LogLk = -532788.357 alpha = 0.621 rescaling lengths by 1.223\n",
      "Total time: 17.38 seconds Unique: 1880/2755 Bad splits: 0/0\n",
      ">>>> RplD_COG0088\n",
      "FastTree Version 2.1.11 SSE3, OpenMP (12 threads)\n",
      "Alignment: standard input\n",
      "Nucleotide distances: Jukes-Cantor Joins: balanced Support: none\n",
      "Start at tree from example_data/RplD_COG0088/est.raxml.bestTree.rooted (no NNI) (no SPR)\n",
      "ML Model: Generalized Time-Reversible, CAT approximation with 20 rate categories\n",
      "Ignored unknown character M (seen 1 times)\n",
      "Ignored unknown character Y (seen 2 times)\n",
      "Total branch-length 198.114 after 0.17 sec\n",
      "      0.19 seconds: ML Lengths 1 of 1788 splits\n",
      "      0.31 seconds: ML Lengths 201 of 1788 splits\n",
      "      0.42 seconds: ML Lengths 401 of 1788 splits\n",
      "      0.54 seconds: ML Lengths 601 of 1788 splits\n",
      "      0.66 seconds: ML Lengths 801 of 1788 splits\n",
      "      0.77 seconds: ML Lengths 1001 of 1788 splits\n",
      "      0.88 seconds: ML Lengths 1201 of 1788 splits\n",
      "      0.99 seconds: ML Lengths 1401 of 1788 splits\n",
      "      1.10 seconds: ML Lengths 1601 of 1788 splits\n",
      "1 rounds ML lengths: LogLk = -503353.989 Max-change 0.2556 Time 1.24\n",
      "      1.24 seconds: Optimizing GTR model, step 1 of 12\n",
      "      1.79 seconds: Optimizing GTR model, step 2 of 12\n",
      "      2.22 seconds: Optimizing GTR model, step 3 of 12\n",
      "      2.64 seconds: Optimizing GTR model, step 4 of 12\n",
      "      3.07 seconds: Optimizing GTR model, step 5 of 12\n",
      "      3.66 seconds: Optimizing GTR model, step 6 of 12\n",
      "      4.30 seconds: Optimizing GTR model, step 7 of 12\n",
      "      4.64 seconds: Optimizing GTR model, step 8 of 12\n",
      "      5.03 seconds: Optimizing GTR model, step 9 of 12\n",
      "      5.41 seconds: Optimizing GTR model, step 10 of 12\n",
      "      5.79 seconds: Optimizing GTR model, step 11 of 12\n",
      "      6.21 seconds: Optimizing GTR model, step 12 of 12\n",
      "GTR Frequencies: 0.2613 0.2515 0.2694 0.2179\n",
      "GTR rates(ac ag at cg ct gt) 1.3986 2.3320 2.0455 1.9842 3.8500 1.0000\n",
      "      6.59 seconds: ML Lengths 1 of 1788 splits\n",
      "      6.74 seconds: ML Lengths 201 of 1788 splits\n",
      "      6.89 seconds: ML Lengths 401 of 1788 splits\n",
      "      7.04 seconds: ML Lengths 601 of 1788 splits\n",
      "      7.19 seconds: ML Lengths 801 of 1788 splits\n",
      "      7.34 seconds: ML Lengths 1001 of 1788 splits\n",
      "      7.49 seconds: ML Lengths 1201 of 1788 splits\n",
      "      7.63 seconds: ML Lengths 1401 of 1788 splits\n",
      "      7.78 seconds: ML Lengths 1601 of 1788 splits\n",
      "      7.97 seconds: Site likelihoods with rate category 1 of 20\n",
      "      8.10 seconds: Site likelihoods with rate category 4 of 20\n",
      "      8.24 seconds: Site likelihoods with rate category 7 of 20\n",
      "      8.37 seconds: Site likelihoods with rate category 10 of 20\n",
      "      8.50 seconds: Site likelihoods with rate category 13 of 20\n",
      "      8.64 seconds: Site likelihoods with rate category 16 of 20\n",
      "      8.77 seconds: Site likelihoods with rate category 19 of 20\n",
      "Switched to using 20 rate categories (CAT approximation)\n",
      "Rate categories were divided by 0.891 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "      8.88 seconds: ML Lengths 1 of 1788 splits\n",
      "      9.04 seconds: ML Lengths 201 of 1788 splits\n",
      "      9.19 seconds: ML Lengths 401 of 1788 splits\n",
      "      9.35 seconds: ML Lengths 601 of 1788 splits\n",
      "      9.52 seconds: ML Lengths 801 of 1788 splits\n",
      "      9.67 seconds: ML Lengths 1001 of 1788 splits\n",
      "      9.82 seconds: ML Lengths 1201 of 1788 splits\n",
      "      9.98 seconds: ML Lengths 1401 of 1788 splits\n",
      "     10.14 seconds: ML Lengths 1601 of 1788 splits\n",
      "2 rounds ML lengths: LogLk = -455760.580 Max-change 0.6034 Time 10.31\n",
      "     10.30 seconds: ML Lengths 1 of 1788 splits\n",
      "     10.45 seconds: ML Lengths 201 of 1788 splits\n",
      "     10.60 seconds: ML Lengths 401 of 1788 splits\n",
      "     10.75 seconds: ML Lengths 601 of 1788 splits\n",
      "     10.90 seconds: ML Lengths 801 of 1788 splits\n",
      "     11.05 seconds: ML Lengths 1001 of 1788 splits\n",
      "     11.20 seconds: ML Lengths 1201 of 1788 splits\n",
      "     11.34 seconds: ML Lengths 1401 of 1788 splits\n",
      "     11.49 seconds: ML Lengths 1601 of 1788 splits\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3 rounds ML lengths: LogLk = -455753.396 Max-change 0.0667 Time 11.66\n",
      "     11.65 seconds: ML Lengths 1 of 1788 splits\n",
      "     11.80 seconds: ML Lengths 201 of 1788 splits\n",
      "     11.94 seconds: ML Lengths 401 of 1788 splits\n",
      "     12.09 seconds: ML Lengths 601 of 1788 splits\n",
      "     12.24 seconds: ML Lengths 801 of 1788 splits\n",
      "     12.39 seconds: ML Lengths 1001 of 1788 splits\n",
      "     12.53 seconds: ML Lengths 1201 of 1788 splits\n",
      "     12.67 seconds: ML Lengths 1401 of 1788 splits\n",
      "     12.82 seconds: ML Lengths 1601 of 1788 splits\n",
      "4 rounds ML lengths: LogLk = -455752.875 Max-change 0.0182 Time 12.98\n",
      "     12.97 seconds: ML Lengths 1 of 1788 splits\n",
      "     13.12 seconds: ML Lengths 201 of 1788 splits\n",
      "     13.26 seconds: ML Lengths 401 of 1788 splits\n",
      "     13.41 seconds: ML Lengths 601 of 1788 splits\n",
      "     13.56 seconds: ML Lengths 801 of 1788 splits\n",
      "     13.70 seconds: ML Lengths 1001 of 1788 splits\n",
      "     13.84 seconds: ML Lengths 1201 of 1788 splits\n",
      "     13.99 seconds: ML Lengths 1401 of 1788 splits\n",
      "     14.14 seconds: ML Lengths 1601 of 1788 splits\n",
      "5 rounds ML lengths: LogLk = -455752.735 Max-change 0.0075 Time 14.29\n",
      "     14.29 seconds: ML Lengths 1 of 1788 splits\n",
      "     14.43 seconds: ML Lengths 201 of 1788 splits\n",
      "     14.58 seconds: ML Lengths 401 of 1788 splits\n",
      "     14.73 seconds: ML Lengths 601 of 1788 splits\n",
      "     14.87 seconds: ML Lengths 801 of 1788 splits\n",
      "     15.02 seconds: ML Lengths 1001 of 1788 splits\n",
      "     15.16 seconds: ML Lengths 1201 of 1788 splits\n",
      "     15.30 seconds: ML Lengths 1401 of 1788 splits\n",
      "     15.45 seconds: ML Lengths 1601 of 1788 splits\n",
      "6 rounds ML lengths: LogLk = -455752.660 Max-change 0.0036 Time 15.61\n",
      "     15.60 seconds: ML Lengths 1 of 1788 splits\n",
      "     15.75 seconds: ML Lengths 201 of 1788 splits\n",
      "     15.89 seconds: ML Lengths 401 of 1788 splits\n",
      "     16.04 seconds: ML Lengths 601 of 1788 splits\n",
      "     16.19 seconds: ML Lengths 801 of 1788 splits\n",
      "     16.33 seconds: ML Lengths 1001 of 1788 splits\n",
      "     16.47 seconds: ML Lengths 1201 of 1788 splits\n",
      "     16.62 seconds: ML Lengths 1401 of 1788 splits\n",
      "     16.76 seconds: ML Lengths 1601 of 1788 splits\n",
      "7 rounds ML lengths: LogLk = -455752.612 Max-change 0.0031 Time 16.92\n",
      "     16.92 seconds: ML Lengths 1 of 1788 splits\n",
      "     17.06 seconds: ML Lengths 201 of 1788 splits\n",
      "     17.21 seconds: ML Lengths 401 of 1788 splits\n",
      "     17.35 seconds: ML Lengths 601 of 1788 splits\n",
      "     17.50 seconds: ML Lengths 801 of 1788 splits\n",
      "     17.65 seconds: ML Lengths 1001 of 1788 splits\n",
      "     17.79 seconds: ML Lengths 1201 of 1788 splits\n",
      "     17.93 seconds: ML Lengths 1401 of 1788 splits\n",
      "     18.08 seconds: ML Lengths 1601 of 1788 splits\n",
      "8 rounds ML lengths: LogLk = -455752.574 Max-change 0.0027 Time 18.23\n",
      "     18.23 seconds: ML Lengths 1 of 1788 splits\n",
      "     18.37 seconds: ML Lengths 201 of 1788 splits\n",
      "     18.52 seconds: ML Lengths 401 of 1788 splits\n",
      "     18.67 seconds: ML Lengths 601 of 1788 splits\n",
      "     18.81 seconds: ML Lengths 801 of 1788 splits\n",
      "     18.96 seconds: ML Lengths 1001 of 1788 splits\n",
      "     19.10 seconds: ML Lengths 1201 of 1788 splits\n",
      "     19.24 seconds: ML Lengths 1401 of 1788 splits\n",
      "     19.39 seconds: ML Lengths 1601 of 1788 splits\n",
      "9 rounds ML lengths: LogLk = -455752.545 Max-change 0.0039 Time 19.55\n",
      "     19.54 seconds: ML Lengths 1 of 1788 splits\n",
      "     19.69 seconds: ML Lengths 201 of 1788 splits\n",
      "     19.83 seconds: ML Lengths 401 of 1788 splits\n",
      "     19.98 seconds: ML Lengths 601 of 1788 splits\n",
      "     20.13 seconds: ML Lengths 801 of 1788 splits\n",
      "     20.27 seconds: ML Lengths 1001 of 1788 splits\n",
      "     20.42 seconds: ML Lengths 1201 of 1788 splits\n",
      "     20.56 seconds: ML Lengths 1401 of 1788 splits\n",
      "     20.71 seconds: ML Lengths 1601 of 1788 splits\n",
      "10 rounds ML lengths: LogLk = -455752.522 Max-change 0.0024 Time 20.86\n",
      "     20.86 seconds: ML Lengths 1 of 1788 splits\n",
      "     21.00 seconds: ML Lengths 201 of 1788 splits\n",
      "     21.15 seconds: ML Lengths 401 of 1788 splits\n",
      "     21.29 seconds: ML Lengths 601 of 1788 splits\n",
      "     21.44 seconds: ML Lengths 801 of 1788 splits\n",
      "     21.59 seconds: ML Lengths 1001 of 1788 splits\n",
      "     21.73 seconds: ML Lengths 1201 of 1788 splits\n",
      "     21.87 seconds: ML Lengths 1401 of 1788 splits\n",
      "     22.02 seconds: ML Lengths 1601 of 1788 splits\n",
      "11 rounds ML lengths: LogLk = -455752.505 Max-change 0.0023 Time 22.17\n",
      "     22.22 seconds: Site likelihoods with rate category 1 of 20\n",
      "     22.36 seconds: Site likelihoods with rate category 4 of 20\n",
      "     22.49 seconds: Site likelihoods with rate category 7 of 20\n",
      "     22.63 seconds: Site likelihoods with rate category 10 of 20\n",
      "     22.77 seconds: Site likelihoods with rate category 13 of 20\n",
      "     22.91 seconds: Site likelihoods with rate category 16 of 20\n",
      "     23.05 seconds: Site likelihoods with rate category 19 of 20\n",
      "Gamma(20) LogLk = -457375.426 alpha = 1.094 rescaling lengths by 1.619\n",
      "Total time: 23.15 seconds Unique: 1790/2731 Bad splits: 0/0\n",
      ">>>> RpsC_COG0092\n",
      "FastTree Version 2.1.11 SSE3, OpenMP (12 threads)\n",
      "Alignment: standard input\n",
      "Nucleotide distances: Jukes-Cantor Joins: balanced Support: none\n",
      "Start at tree from example_data/RpsC_COG0092/est.raxml.bestTree.rooted (no NNI) (no SPR)\n",
      "ML Model: Generalized Time-Reversible, CAT approximation with 20 rate categories\n",
      "Ignored unknown character R (seen 1 times)\n",
      "Total branch-length 157.953 after 0.39 sec\n",
      "      0.42 seconds: ML Lengths 1 of 1811 splits\n",
      "      0.57 seconds: ML Lengths 101 of 1811 splits\n",
      "      0.72 seconds: ML Lengths 201 of 1811 splits\n",
      "      0.87 seconds: ML Lengths 301 of 1811 splits\n",
      "      1.02 seconds: ML Lengths 401 of 1811 splits\n",
      "      1.17 seconds: ML Lengths 501 of 1811 splits\n",
      "      1.33 seconds: ML Lengths 601 of 1811 splits\n",
      "      1.49 seconds: ML Lengths 701 of 1811 splits\n",
      "      1.64 seconds: ML Lengths 801 of 1811 splits\n",
      "      1.79 seconds: ML Lengths 901 of 1811 splits\n",
      "      1.93 seconds: ML Lengths 1001 of 1811 splits\n",
      "      2.07 seconds: ML Lengths 1101 of 1811 splits\n",
      "      2.22 seconds: ML Lengths 1201 of 1811 splits\n",
      "      2.38 seconds: ML Lengths 1301 of 1811 splits\n",
      "      2.53 seconds: ML Lengths 1401 of 1811 splits\n",
      "      2.68 seconds: ML Lengths 1501 of 1811 splits\n",
      "      2.83 seconds: ML Lengths 1601 of 1811 splits\n",
      "      2.99 seconds: ML Lengths 1701 of 1811 splits\n",
      "      3.15 seconds: ML Lengths 1801 of 1811 splits\n",
      "1 rounds ML lengths: LogLk = -512675.126 Max-change 0.2292 Time 3.24\n",
      "      4.13 seconds: Optimizing GTR model, step 2 of 12\n",
      "      4.75 seconds: Optimizing GTR model, step 3 of 12\n",
      "      5.43 seconds: Optimizing GTR model, step 4 of 12\n",
      "      6.26 seconds: Optimizing GTR model, step 5 of 12\n",
      "      7.15 seconds: Optimizing GTR model, step 6 of 12\n",
      "      8.31 seconds: Optimizing GTR model, step 7 of 12\n",
      "      8.93 seconds: Optimizing GTR model, step 8 of 12\n",
      "      9.55 seconds: Optimizing GTR model, step 9 of 12\n",
      "     10.23 seconds: Optimizing GTR model, step 10 of 12\n",
      "     10.85 seconds: Optimizing GTR model, step 11 of 12\n",
      "     11.54 seconds: Optimizing GTR model, step 12 of 12\n",
      "GTR Frequencies: 0.2631 0.2485 0.2822 0.2063\n",
      "GTR rates(ac ag at cg ct gt) 1.5677 2.5258 2.4485 2.0320 5.3859 1.0000\n",
      "     12.20 seconds: ML Lengths 1 of 1811 splits\n",
      "     12.39 seconds: ML Lengths 101 of 1811 splits\n",
      "     12.58 seconds: ML Lengths 201 of 1811 splits\n",
      "     12.76 seconds: ML Lengths 301 of 1811 splits\n",
      "     12.94 seconds: ML Lengths 401 of 1811 splits\n",
      "     13.13 seconds: ML Lengths 501 of 1811 splits\n",
      "     13.32 seconds: ML Lengths 601 of 1811 splits\n",
      "     13.51 seconds: ML Lengths 701 of 1811 splits\n",
      "     13.69 seconds: ML Lengths 801 of 1811 splits\n",
      "     13.88 seconds: ML Lengths 901 of 1811 splits\n",
      "     14.06 seconds: ML Lengths 1001 of 1811 splits\n",
      "     14.24 seconds: ML Lengths 1101 of 1811 splits\n",
      "     14.42 seconds: ML Lengths 1201 of 1811 splits\n",
      "     14.61 seconds: ML Lengths 1301 of 1811 splits\n",
      "     14.80 seconds: ML Lengths 1401 of 1811 splits\n",
      "     14.98 seconds: ML Lengths 1501 of 1811 splits\n",
      "     15.16 seconds: ML Lengths 1601 of 1811 splits\n",
      "     15.35 seconds: ML Lengths 1701 of 1811 splits\n",
      "     15.54 seconds: ML Lengths 1801 of 1811 splits\n",
      "     15.71 seconds: Site likelihoods with rate category 2 of 20\n",
      "     15.85 seconds: Site likelihoods with rate category 4 of 20\n",
      "     16.00 seconds: Site likelihoods with rate category 6 of 20\n",
      "     16.14 seconds: Site likelihoods with rate category 8 of 20\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     16.29 seconds: Site likelihoods with rate category 10 of 20\n",
      "     16.43 seconds: Site likelihoods with rate category 12 of 20\n",
      "     16.58 seconds: Site likelihoods with rate category 14 of 20\n",
      "     16.72 seconds: Site likelihoods with rate category 16 of 20\n",
      "     16.87 seconds: Site likelihoods with rate category 18 of 20\n",
      "     17.01 seconds: Site likelihoods with rate category 20 of 20\n",
      "Switched to using 20 rate categories (CAT approximation)\n",
      "Rate categories were divided by 0.909 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "     17.12 seconds: ML Lengths 1 of 1811 splits\n",
      "     17.31 seconds: ML Lengths 101 of 1811 splits\n",
      "     17.50 seconds: ML Lengths 201 of 1811 splits\n",
      "     17.69 seconds: ML Lengths 301 of 1811 splits\n",
      "     17.88 seconds: ML Lengths 401 of 1811 splits\n",
      "     18.07 seconds: ML Lengths 501 of 1811 splits\n",
      "     18.28 seconds: ML Lengths 601 of 1811 splits\n",
      "     18.47 seconds: ML Lengths 701 of 1811 splits\n",
      "     18.67 seconds: ML Lengths 801 of 1811 splits\n",
      "     18.86 seconds: ML Lengths 901 of 1811 splits\n",
      "     19.04 seconds: ML Lengths 1001 of 1811 splits\n",
      "     19.22 seconds: ML Lengths 1101 of 1811 splits\n",
      "     19.41 seconds: ML Lengths 1201 of 1811 splits\n",
      "     19.61 seconds: ML Lengths 1301 of 1811 splits\n",
      "     19.80 seconds: ML Lengths 1401 of 1811 splits\n",
      "     19.99 seconds: ML Lengths 1501 of 1811 splits\n",
      "     20.18 seconds: ML Lengths 1601 of 1811 splits\n",
      "     20.38 seconds: ML Lengths 1701 of 1811 splits\n",
      "     20.58 seconds: ML Lengths 1801 of 1811 splits\n",
      "2 rounds ML lengths: LogLk = -447811.218 Max-change 1.2107 Time 20.63\n",
      "     20.81 seconds: ML Lengths 101 of 1811 splits\n",
      "     20.99 seconds: ML Lengths 201 of 1811 splits\n",
      "     21.17 seconds: ML Lengths 301 of 1811 splits\n",
      "     21.35 seconds: ML Lengths 401 of 1811 splits\n",
      "     21.53 seconds: ML Lengths 501 of 1811 splits\n",
      "     21.72 seconds: ML Lengths 601 of 1811 splits\n",
      "     21.90 seconds: ML Lengths 701 of 1811 splits\n",
      "     22.08 seconds: ML Lengths 801 of 1811 splits\n",
      "     22.27 seconds: ML Lengths 901 of 1811 splits\n",
      "     22.44 seconds: ML Lengths 1001 of 1811 splits\n",
      "     22.61 seconds: ML Lengths 1101 of 1811 splits\n",
      "     22.79 seconds: ML Lengths 1201 of 1811 splits\n",
      "     22.98 seconds: ML Lengths 1301 of 1811 splits\n",
      "     23.16 seconds: ML Lengths 1401 of 1811 splits\n",
      "     23.34 seconds: ML Lengths 1501 of 1811 splits\n",
      "     23.52 seconds: ML Lengths 1601 of 1811 splits\n",
      "     23.71 seconds: ML Lengths 1701 of 1811 splits\n",
      "     23.89 seconds: ML Lengths 1801 of 1811 splits\n",
      "3 rounds ML lengths: LogLk = -447799.539 Max-change 0.1344 Time 23.94\n",
      "     24.12 seconds: ML Lengths 101 of 1811 splits\n",
      "     24.30 seconds: ML Lengths 201 of 1811 splits\n",
      "     24.47 seconds: ML Lengths 301 of 1811 splits\n",
      "     24.65 seconds: ML Lengths 401 of 1811 splits\n",
      "     24.83 seconds: ML Lengths 501 of 1811 splits\n",
      "     25.01 seconds: ML Lengths 601 of 1811 splits\n",
      "     25.19 seconds: ML Lengths 701 of 1811 splits\n",
      "     25.37 seconds: ML Lengths 801 of 1811 splits\n",
      "     25.55 seconds: ML Lengths 901 of 1811 splits\n",
      "     25.72 seconds: ML Lengths 1001 of 1811 splits\n",
      "     25.89 seconds: ML Lengths 1101 of 1811 splits\n",
      "     26.07 seconds: ML Lengths 1201 of 1811 splits\n",
      "     26.25 seconds: ML Lengths 1301 of 1811 splits\n",
      "     26.43 seconds: ML Lengths 1401 of 1811 splits\n",
      "     26.61 seconds: ML Lengths 1501 of 1811 splits\n",
      "     26.79 seconds: ML Lengths 1601 of 1811 splits\n",
      "     26.97 seconds: ML Lengths 1701 of 1811 splits\n",
      "     27.15 seconds: ML Lengths 1801 of 1811 splits\n",
      "4 rounds ML lengths: LogLk = -447798.480 Max-change 0.0380 Time 27.20\n",
      "     27.37 seconds: ML Lengths 101 of 1811 splits\n",
      "     27.55 seconds: ML Lengths 201 of 1811 splits\n",
      "     27.73 seconds: ML Lengths 301 of 1811 splits\n",
      "     27.90 seconds: ML Lengths 401 of 1811 splits\n",
      "     28.08 seconds: ML Lengths 501 of 1811 splits\n",
      "     28.27 seconds: ML Lengths 601 of 1811 splits\n",
      "     28.44 seconds: ML Lengths 701 of 1811 splits\n",
      "     28.62 seconds: ML Lengths 801 of 1811 splits\n",
      "     28.80 seconds: ML Lengths 901 of 1811 splits\n",
      "     28.97 seconds: ML Lengths 1001 of 1811 splits\n",
      "     29.14 seconds: ML Lengths 1101 of 1811 splits\n",
      "     29.32 seconds: ML Lengths 1201 of 1811 splits\n",
      "     29.50 seconds: ML Lengths 1301 of 1811 splits\n",
      "     29.68 seconds: ML Lengths 1401 of 1811 splits\n",
      "     29.85 seconds: ML Lengths 1501 of 1811 splits\n",
      "     30.03 seconds: ML Lengths 1601 of 1811 splits\n",
      "     30.21 seconds: ML Lengths 1701 of 1811 splits\n",
      "     30.39 seconds: ML Lengths 1801 of 1811 splits\n",
      "5 rounds ML lengths: LogLk = -447798.300 Max-change 0.0113 Time 30.44\n",
      "     30.61 seconds: ML Lengths 101 of 1811 splits\n",
      "     30.79 seconds: ML Lengths 201 of 1811 splits\n",
      "     30.97 seconds: ML Lengths 301 of 1811 splits\n",
      "     31.14 seconds: ML Lengths 401 of 1811 splits\n",
      "     31.32 seconds: ML Lengths 501 of 1811 splits\n",
      "     31.50 seconds: ML Lengths 601 of 1811 splits\n",
      "     31.68 seconds: ML Lengths 701 of 1811 splits\n",
      "     31.86 seconds: ML Lengths 801 of 1811 splits\n",
      "     32.04 seconds: ML Lengths 901 of 1811 splits\n",
      "     32.21 seconds: ML Lengths 1001 of 1811 splits\n",
      "     32.38 seconds: ML Lengths 1101 of 1811 splits\n",
      "     32.56 seconds: ML Lengths 1201 of 1811 splits\n",
      "     32.74 seconds: ML Lengths 1301 of 1811 splits\n",
      "     32.91 seconds: ML Lengths 1401 of 1811 splits\n",
      "     33.09 seconds: ML Lengths 1501 of 1811 splits\n",
      "     33.27 seconds: ML Lengths 1601 of 1811 splits\n",
      "     33.45 seconds: ML Lengths 1701 of 1811 splits\n",
      "     33.63 seconds: ML Lengths 1801 of 1811 splits\n",
      "6 rounds ML lengths: LogLk = -447798.225 Max-change 0.0034 Time 33.68\n",
      "     33.85 seconds: ML Lengths 101 of 1811 splits\n",
      "     34.03 seconds: ML Lengths 201 of 1811 splits\n",
      "     34.20 seconds: ML Lengths 301 of 1811 splits\n",
      "     34.38 seconds: ML Lengths 401 of 1811 splits\n",
      "     34.56 seconds: ML Lengths 501 of 1811 splits\n",
      "     34.74 seconds: ML Lengths 601 of 1811 splits\n",
      "     34.92 seconds: ML Lengths 701 of 1811 splits\n",
      "     35.10 seconds: ML Lengths 801 of 1811 splits\n",
      "     35.27 seconds: ML Lengths 901 of 1811 splits\n",
      "     35.45 seconds: ML Lengths 1001 of 1811 splits\n",
      "     35.62 seconds: ML Lengths 1101 of 1811 splits\n",
      "     35.79 seconds: ML Lengths 1201 of 1811 splits\n",
      "     35.97 seconds: ML Lengths 1301 of 1811 splits\n",
      "     36.15 seconds: ML Lengths 1401 of 1811 splits\n",
      "     36.33 seconds: ML Lengths 1501 of 1811 splits\n",
      "     36.50 seconds: ML Lengths 1601 of 1811 splits\n",
      "     36.68 seconds: ML Lengths 1701 of 1811 splits\n",
      "     36.86 seconds: ML Lengths 1801 of 1811 splits\n",
      "7 rounds ML lengths: LogLk = -447798.207 Max-change 0.0022 Time 36.91\n",
      "     37.08 seconds: ML Lengths 101 of 1811 splits\n",
      "     37.26 seconds: ML Lengths 201 of 1811 splits\n",
      "     37.44 seconds: ML Lengths 301 of 1811 splits\n",
      "     37.61 seconds: ML Lengths 401 of 1811 splits\n",
      "     37.79 seconds: ML Lengths 501 of 1811 splits\n",
      "     37.97 seconds: ML Lengths 601 of 1811 splits\n",
      "     38.15 seconds: ML Lengths 701 of 1811 splits\n",
      "     38.33 seconds: ML Lengths 801 of 1811 splits\n",
      "     38.51 seconds: ML Lengths 901 of 1811 splits\n",
      "     38.68 seconds: ML Lengths 1001 of 1811 splits\n",
      "     38.85 seconds: ML Lengths 1101 of 1811 splits\n",
      "     39.02 seconds: ML Lengths 1201 of 1811 splits\n",
      "     39.20 seconds: ML Lengths 1301 of 1811 splits\n",
      "     39.38 seconds: ML Lengths 1401 of 1811 splits\n",
      "     39.56 seconds: ML Lengths 1501 of 1811 splits\n",
      "     39.73 seconds: ML Lengths 1601 of 1811 splits\n",
      "     39.91 seconds: ML Lengths 1701 of 1811 splits\n",
      "     40.09 seconds: ML Lengths 1801 of 1811 splits\n",
      "8 rounds ML lengths: LogLk = -447798.201 Max-change 0.0013 Time 40.14\n",
      "     40.31 seconds: ML Lengths 101 of 1811 splits\n",
      "     40.49 seconds: ML Lengths 201 of 1811 splits\n",
      "     40.67 seconds: ML Lengths 301 of 1811 splits\n",
      "     40.84 seconds: ML Lengths 401 of 1811 splits\n",
      "     41.02 seconds: ML Lengths 501 of 1811 splits\n",
      "     41.20 seconds: ML Lengths 601 of 1811 splits\n",
      "     41.38 seconds: ML Lengths 701 of 1811 splits\n",
      "     41.56 seconds: ML Lengths 801 of 1811 splits\n",
      "     41.74 seconds: ML Lengths 901 of 1811 splits\n",
      "     41.91 seconds: ML Lengths 1001 of 1811 splits\n",
      "     42.08 seconds: ML Lengths 1101 of 1811 splits\n",
      "     42.25 seconds: ML Lengths 1201 of 1811 splits\n",
      "     42.43 seconds: ML Lengths 1301 of 1811 splits\n",
      "     42.61 seconds: ML Lengths 1401 of 1811 splits\n",
      "     42.79 seconds: ML Lengths 1501 of 1811 splits\n",
      "     42.97 seconds: ML Lengths 1601 of 1811 splits\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     43.14 seconds: ML Lengths 1701 of 1811 splits\n",
      "     43.32 seconds: ML Lengths 1801 of 1811 splits\n",
      "9 rounds ML lengths: LogLk = -447798.199 Max-change 0.0011 Time 43.37\n",
      "     43.55 seconds: ML Lengths 101 of 1811 splits\n",
      "     43.72 seconds: ML Lengths 201 of 1811 splits\n",
      "     43.90 seconds: ML Lengths 301 of 1811 splits\n",
      "     44.07 seconds: ML Lengths 401 of 1811 splits\n",
      "     44.25 seconds: ML Lengths 501 of 1811 splits\n",
      "     44.43 seconds: ML Lengths 601 of 1811 splits\n",
      "     44.61 seconds: ML Lengths 701 of 1811 splits\n",
      "     44.79 seconds: ML Lengths 801 of 1811 splits\n",
      "     44.97 seconds: ML Lengths 901 of 1811 splits\n",
      "     45.14 seconds: ML Lengths 1001 of 1811 splits\n",
      "     45.31 seconds: ML Lengths 1101 of 1811 splits\n",
      "     45.49 seconds: ML Lengths 1201 of 1811 splits\n",
      "     45.66 seconds: ML Lengths 1301 of 1811 splits\n",
      "     45.84 seconds: ML Lengths 1401 of 1811 splits\n",
      "     46.02 seconds: ML Lengths 1501 of 1811 splits\n",
      "     46.19 seconds: ML Lengths 1601 of 1811 splits\n",
      "     46.37 seconds: ML Lengths 1701 of 1811 splits\n",
      "     46.55 seconds: ML Lengths 1801 of 1811 splits\n",
      "10 rounds ML lengths: LogLk = -447798.198 Max-change 0.0005 (converged) Time 46.60\n",
      "     46.67 seconds: Site likelihoods with rate category 1 of 20\n",
      "     46.82 seconds: Site likelihoods with rate category 3 of 20\n",
      "     46.97 seconds: Site likelihoods with rate category 5 of 20\n",
      "     47.12 seconds: Site likelihoods with rate category 7 of 20\n",
      "     47.26 seconds: Site likelihoods with rate category 9 of 20\n",
      "     47.41 seconds: Site likelihoods with rate category 11 of 20\n",
      "     47.56 seconds: Site likelihoods with rate category 13 of 20\n",
      "     47.71 seconds: Site likelihoods with rate category 15 of 20\n",
      "     47.86 seconds: Site likelihoods with rate category 17 of 20\n",
      "     48.00 seconds: Site likelihoods with rate category 19 of 20\n",
      "     48.13 seconds: Optimizing alpha round 1\n",
      "Gamma(20) LogLk = -449926.695 alpha = 0.946 rescaling lengths by 2.366\n",
      "Total time: 48.20 seconds Unique: 1813/2731 Bad splits: 0/0\n",
      ">>>> RpsM_COG0099\n",
      "FastTree Version 2.1.11 SSE3, OpenMP (12 threads)\n",
      "Alignment: standard input\n",
      "Nucleotide distances: Jukes-Cantor Joins: balanced Support: none\n",
      "Start at tree from example_data/RpsM_COG0099/est.raxml.bestTree.rooted (no NNI) (no SPR)\n",
      "ML Model: Generalized Time-Reversible, CAT approximation with 20 rate categories\n",
      "Total branch-length 153.393 after 0.06 sec\n",
      "      0.10 seconds: ML Lengths 201 of 1702 splits\n",
      "      0.21 seconds: ML Lengths 901 of 1702 splits\n",
      "      0.32 seconds: ML Lengths 1601 of 1702 splits\n",
      "1 rounds ML lengths: LogLk = -253078.872 Max-change 0.2135 Time 0.35\n",
      "      0.60 seconds: Optimizing GTR model, step 2 of 12\n",
      "      0.78 seconds: Optimizing GTR model, step 3 of 12\n",
      "      1.06 seconds: Optimizing GTR model, step 4 of 12\n",
      "      1.35 seconds: Optimizing GTR model, step 5 of 12\n",
      "      1.59 seconds: Optimizing GTR model, step 6 of 12\n",
      "      1.88 seconds: Optimizing GTR model, step 7 of 12\n",
      "      2.06 seconds: Optimizing GTR model, step 8 of 12\n",
      "      2.24 seconds: Optimizing GTR model, step 9 of 12\n",
      "      2.41 seconds: Optimizing GTR model, step 10 of 12\n",
      "      2.59 seconds: Optimizing GTR model, step 11 of 12\n",
      "      2.79 seconds: Optimizing GTR model, step 12 of 12\n",
      "GTR Frequencies: 0.2741 0.2499 0.2693 0.2068\n",
      "GTR rates(ac ag at cg ct gt) 1.4094 2.3583 2.2884 1.6215 5.4830 1.0000\n",
      "      2.98 seconds: ML Lengths 1 of 1702 splits\n",
      "      3.09 seconds: ML Lengths 501 of 1702 splits\n",
      "      3.20 seconds: ML Lengths 1001 of 1702 splits\n",
      "      3.32 seconds: ML Lengths 1501 of 1702 splits\n",
      "      3.43 seconds: Site likelihoods with rate category 3 of 20\n",
      "      3.53 seconds: Site likelihoods with rate category 8 of 20\n",
      "      3.63 seconds: Site likelihoods with rate category 13 of 20\n",
      "      3.73 seconds: Site likelihoods with rate category 18 of 20\n",
      "Switched to using 20 rate categories (CAT approximation)\n",
      "Rate categories were divided by 0.985 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "      3.86 seconds: ML Lengths 201 of 1702 splits\n",
      "      3.96 seconds: ML Lengths 601 of 1702 splits\n",
      "      4.08 seconds: ML Lengths 1101 of 1702 splits\n",
      "      4.18 seconds: ML Lengths 1501 of 1702 splits\n",
      "2 rounds ML lengths: LogLk = -218165.779 Max-change 0.8251 Time 4.24\n",
      "      4.29 seconds: ML Lengths 201 of 1702 splits\n",
      "      4.41 seconds: ML Lengths 701 of 1702 splits\n",
      "      4.53 seconds: ML Lengths 1201 of 1702 splits\n",
      "      4.65 seconds: ML Lengths 1701 of 1702 splits\n",
      "3 rounds ML lengths: LogLk = -218148.471 Max-change 0.1159 Time 4.66\n",
      "      4.75 seconds: ML Lengths 401 of 1702 splits\n",
      "      4.87 seconds: ML Lengths 901 of 1702 splits\n",
      "      4.99 seconds: ML Lengths 1401 of 1702 splits\n",
      "4 rounds ML lengths: LogLk = -218147.094 Max-change 0.0233 Time 5.07\n",
      "      5.09 seconds: ML Lengths 101 of 1702 splits\n",
      "      5.21 seconds: ML Lengths 601 of 1702 splits\n",
      "      5.33 seconds: ML Lengths 1101 of 1702 splits\n",
      "      5.45 seconds: ML Lengths 1601 of 1702 splits\n",
      "5 rounds ML lengths: LogLk = -218146.929 Max-change 0.0055 Time 5.48\n",
      "      5.55 seconds: ML Lengths 301 of 1702 splits\n",
      "      5.66 seconds: ML Lengths 801 of 1702 splits\n",
      "      5.78 seconds: ML Lengths 1301 of 1702 splits\n",
      "6 rounds ML lengths: LogLk = -218146.889 Max-change 0.0018 Time 5.89\n",
      "      5.88 seconds: ML Lengths 1 of 1702 splits\n",
      "      6.00 seconds: ML Lengths 501 of 1702 splits\n",
      "      6.12 seconds: ML Lengths 1001 of 1702 splits\n",
      "      6.23 seconds: ML Lengths 1501 of 1702 splits\n",
      "7 rounds ML lengths: LogLk = -218146.871 Max-change 0.0014 Time 6.29\n",
      "      6.34 seconds: ML Lengths 201 of 1702 splits\n",
      "      6.45 seconds: ML Lengths 701 of 1702 splits\n",
      "      6.57 seconds: ML Lengths 1201 of 1702 splits\n",
      "      6.69 seconds: ML Lengths 1701 of 1702 splits\n",
      "8 rounds ML lengths: LogLk = -218146.862 Max-change 0.0012 Time 6.70\n",
      "      6.79 seconds: ML Lengths 401 of 1702 splits\n",
      "      6.90 seconds: ML Lengths 901 of 1702 splits\n",
      "      7.02 seconds: ML Lengths 1401 of 1702 splits\n",
      "9 rounds ML lengths: LogLk = -218146.856 Max-change 0.0007 (converged) Time 7.10\n",
      "      7.14 seconds: Site likelihoods with rate category 2 of 20\n",
      "      7.25 seconds: Site likelihoods with rate category 7 of 20\n",
      "      7.36 seconds: Site likelihoods with rate category 12 of 20\n",
      "      7.47 seconds: Site likelihoods with rate category 17 of 20\n",
      "Gamma(20) LogLk = -219052.669 alpha = 0.616 rescaling lengths by 1.180\n",
      "Total time: 7.57 seconds Unique: 1704/2769 Bad splits: 0/0\n"
     ]
    }
   ],
   "source": [
    "for gene in genes:\n",
    "    print('>>>>', gene)\n",
    "    indir = os.path.join(workdir, gene)\n",
    "    \n",
    "    # use the original alignment and raxml-ng tree to re-estimate FastTree-2 numeric parameters\n",
    "    rooted_tree_path = os.path.join(indir, 'est.raxml.bestTree.rooted')\n",
    "    alignment_path = os.path.join(indir, 'est.aln.nuc.fasta.gz')\n",
    "    ft_log_path = os.path.join(indir, 'est.fasttree.log')\n",
    "    ft_tree_path = os.path.join(indir, 'est.fasttree.tre')\n",
    "    \n",
    "    # enable multi-thread FastTree-2 (using binary FastTreeMP)\n",
    "    os.system(f\"export OMP_NUM_THREADS={num_threads}\")\n",
    "    \n",
    "    # nucleotide alignment\n",
    "    cmd = ['FastTreeMP', '-nosupport', '-gtr', '-gamma', '-nt']\n",
    "    cmd.extend([\n",
    "        '-mllen', '-nome',\n",
    "        '-log', ft_log_path, '-intree', rooted_tree_path,\n",
    "    ])\n",
    "    \n",
    "    if alignment_path.split('.')[-1] == 'gz':\n",
    "        cmd = ['gzip', '-d', alignment_path, '-c', '|'] + cmd\n",
    "    else:\n",
    "        cmd = ['cat', alignment_path, '|'] + cmd\n",
    "    out_fptr = open(ft_tree_path, 'w')\n",
    "    p = Popen(' '.join(cmd), stdout=out_fptr, stderr=PIPE, shell=True, text=True)#, stdin=in_fptr)\n",
    "    for line in p.stderr:\n",
    "        print(line.replace('\\n', ''))\n",
    "    out_fptr.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3133cfab",
   "metadata": {},
   "source": [
    "# Final step: create the reference package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "83a3649b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 03/05/2025 14:30:03\n",
      "New DB name:   /home/chengzes/Desktop/Research/softwares/TIPP3/refpkg_scripts/custom_tipp_refpkg/markers-v4/blast/alignment.fasta.db\n",
      "New DB title:  custom_tipp_refpkg/markers-v4/blast/alignment.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /home/chengzes/Desktop/Research/softwares/TIPP3/refpkg_scripts/custom_tipp_refpkg/markers-v4/blast/alignment.fasta.db\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 13735 sequences in 0.123385 seconds.\n",
      "\n",
      "\n",
      "total unique taxids: 10960\n",
      "Aggregating all sequences to create a BLAST database...\n",
      "Running makeblastdb...\n",
      "done creating refpkg: RpsM_COG0099\n",
      "done creating refpkg: RplA_COG0081\n",
      "done creating refpkg: RplB_COG0090\n",
      "done creating refpkg: RplD_COG0088\n",
      "done creating refpkg: RpsC_COG0092\n",
      "done computing num sequences: RplA_COG0081\n",
      "done computing num sequences: RplB_COG0090\n",
      "done computing num sequences: RplD_COG0088\n",
      "done computing num sequences: RpsM_COG0099\n",
      "done computing num sequences: RpsC_COG0092\n"
     ]
    }
   ],
   "source": [
    "# run the existing script create_tipp_refpkg.py\n",
    "db_path = os.path.join(workdir, 'ncbi_taxonomy.db')\n",
    "\n",
    "cmd = f\"python create_tipp_refpkg.py {workdir} {db_path} {refpkg_dir} {num_threads}\"\n",
    "p = Popen(cmd, stdout=PIPE, stderr=STDOUT, text=True, shell=True)\n",
    "for line in p.stdout:\n",
    "    print(line.replace('\\n', ''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f78ff556",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
