{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eb5d9dfd",
   "metadata": {},
   "source": [
    "# Overview\n",
    "This file contains an example workflow to create a customized TIPP reference package, using some example genes included under `example_data/` with their necessary files:\n",
    "1. the alignment,\n",
    "2. the taxid of sequences,\n",
    "3. the mapping between each sequence and its taxid.\n",
    "\n",
    "----------------------\n",
    "# Required Packages\n",
    "```\n",
    "taxtastic>=0.10.0\n",
    "dendropy>=4.5.2\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b47d2ea0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import os, sys\n",
    "from subprocess import Popen, PIPE, STDOUT\n",
    "\n",
    "# number of threads to use for raxml-ng and fasttree-2\n",
    "num_threads = 4\n",
    "\n",
    "# initialize work directory\n",
    "workdir = 'example_data'\n",
    "\n",
    "# where to write the TIPP reference package to\n",
    "refpkg_dir = 'custom_tipp_refpkg'\n",
    "\n",
    "# initialize the list of genes to use, which are assumed to be under the working directory\n",
    "genes = ['RplA_COG0081', 'RplB_COG0090', 'RplD_COG0088', 'RpsC_COG0092', 'RpsM_COG0099']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bc6b2d8",
   "metadata": {},
   "source": [
    "# Needed files\n",
    "In this example, we are only using one gene to create our customized TIPP reference package, included under `example_data/RplO_COG0200`. Remember that there are 4 files that we need to prepare (one for global use, and three for each gene):\n",
    "1. `ncbi_taxonomy.db  - (global) NCBI taxonomy file created with taxtastic`\n",
    "2. `est.aln.nuc.fasta[.gz] - (each gene) alignment file (can be gzipped)`\n",
    "3. `species.txt       - (each gene) taxids of sequences in alignment, in the same order as in the alignment`\n",
    "4. `species.mapping   - (each gene) mapping between sequences and taxids, in the same order as in the alignment`\n",
    "\n",
    "The three files for the example gene are included, and the NCBI taxonomy database can be built below."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a65e8e05",
   "metadata": {},
   "source": [
    "# Step 0: NCBI taxonomy database - `ncbi_taxonomy.db`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1b0fd8fa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# download the latest NCBI taxdmp.zip\n",
    "url = \"https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdmp.zip\"\n",
    "\n",
    "# run taxtastic to create a database locally\n",
    "if not os.path.exists(f\"{workdir}/ncbi_taxonomy.db\"):\n",
    "    cmd = f\"wget -P {workdir} {url}\"\n",
    "    os.system(cmd)\n",
    "\n",
    "    cmd = f\"taxit new_database -z {workdir}/taxdmp.zip\"\n",
    "    os.system(cmd)\n",
    "\n",
    "    # move the created ncbi_taxonomy.db file to workdir\n",
    "    os.system(f\"mv ncbi_taxonomy.db {workdir}/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cb805e7",
   "metadata": {},
   "source": [
    "# Step 1: Update species taxids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "104ea3ec",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "db_path = os.path.join(workdir, 'ncbi_taxonomy.db')\n",
    "\n",
    "# two files to update\n",
    "old2new_map = {'species.txt': 'species.updated.txt',\n",
    "               'species.mapping': 'species.updated.mapping',\n",
    "              }\n",
    "\n",
    "for gene in genes:\n",
    "    indir = os.path.join(workdir, gene)\n",
    "\n",
    "    for infile, outfile in old2new_map.items():\n",
    "        inpath = os.path.join(indir, infile)\n",
    "        outpath = os.path.join(indir, outfile)\n",
    "        # run taxit update_taxids\n",
    "        cmd = f\"taxit update_taxids {inpath} {db_path} -o {outpath}\"\n",
    "        os.system(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76e31710",
   "metadata": {},
   "source": [
    "# Step 2: Get taxonomy table from the included species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b05f435d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "db_path = os.path.join(workdir, 'ncbi_taxonomy.db')\n",
    "\n",
    "for gene in genes:\n",
    "    indir = os.path.join(workdir, gene)\n",
    "    \n",
    "    # get taxonomy.table with taxit\n",
    "    species_path = os.path.join(indir, 'species.updated.txt')\n",
    "    taxtable_path = os.path.join(indir, 'taxonomy.table')\n",
    "    cmd = f\"taxit taxtable {db_path} -i {species_path} -o {taxtable_path}\"\n",
    "    os.system(cmd)\n",
    "    \n",
    "    # get a cleaned version of the taxonomy table by removing redundant columns\n",
    "    alltaxon_path = os.path.join(indir, 'all_taxon.taxonomy')\n",
    "    cmd = f\"python clean_taxonomy_table.py {taxtable_path} {alltaxon_path}\"\n",
    "    os.system(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f976fc8f",
   "metadata": {},
   "source": [
    "# Step 3: Initialize taxonomy tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7eb43c4f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>>> RplA_COG0081\n",
      "Child taxids at root (131567): 2,2157\n",
      ">>>> RplB_COG0090\n",
      "Child taxids at root (131567): 2\n",
      "WARNING: node (taxid 1275) is internal but have assigned sequences!\n",
      "\tRe-assigning sequences with taxid 136273 to 1275\n",
      "Reassigning taxids to sequences, backing up example_data/RplB_COG0090/species.updated.txt and example_data/RplB_COG0090/species.updated.mapping ...\n",
      "GCF_001483755.1_NZ_LQBK01000002.1_cds_WP_058872530.1_238 taxid reassigned from 136273 to 1275\n",
      "Reassigned taxids for 1/5510 sequences\n",
      ">>>> RplD_COG0088\n",
      "Child taxids at root (131567): 2\n",
      "WARNING: node (taxid 561879) is internal but have assigned sequences!\n",
      "\tRe-assigning sequences with taxid 1178541 to 561879\n",
      "WARNING: node (taxid 1421) is internal but have assigned sequences!\n",
      "\tRe-assigning sequences with taxid 444177 to 1421\n",
      "Reassigning taxids to sequences, backing up example_data/RplD_COG0088/species.updated.txt and example_data/RplD_COG0088/species.updated.mapping ...\n",
      "GCF_000017965.1_NC_010382.1_cds_WP_012296062.1_4284 taxid reassigned from 444177 to 1421\n",
      "GCF_000691165.1_NZ_ASJD01000012.1_cds_WP_024425598.1_3451 taxid reassigned from 1178541 to 561879\n",
      "Reassigned taxids for 2/5462 sequences\n",
      ">>>> RpsC_COG0092\n",
      "Child taxids at root (131567): 2,2157\n",
      ">>>> RpsM_COG0099\n",
      "Child taxids at root (131567): 2\n"
     ]
    }
   ],
   "source": [
    "for gene in genes:\n",
    "    print('>>>>', gene)\n",
    "    indir = os.path.join(workdir, gene)\n",
    "\n",
    "    # run script build_taxonomic_tree.py\n",
    "    # NOTE: both species.updated.txt and species.updated.mapping may be updated\n",
    "    #       if there are unifurication detected in the taxonomy tree.\n",
    "    #       e.g., if we have the following taxids: 57493 --> 1275 --> 136273\n",
    "    #       then, both nodes 1275 and 136273 have sequences assigne to them.\n",
    "    #       However, since node 1275 only has one child, it will be \"contracted\" when\n",
    "    #       eliminating unifurication. Thus, sequences assigned to it may not appear\n",
    "    #       in the unrefined.taxonomy.renamed file and leads to error when running RAxML-ng\n",
    "    #\n",
    "    #       This is a rare event, and the current solution is to modify nodes so that if we find\n",
    "    #       a unifuricating node with taxid and sequences are assigned to it, we will contract all\n",
    "    #       branches below this node and re-assign the sequences belonging to leaves of this subtree\n",
    "    #       (rooted at the unifuricating node) to this node.\n",
    "    #       Then, both species.updated.txt and species.updated.mapping would be updated to reflect\n",
    "    #       the reassignment of taxids (to the sequences that are affected).\n",
    "    species_path = os.path.join(indir, 'species.updated.txt')\n",
    "    mapping_path = os.path.join(indir, 'species.updated.mapping')\n",
    "    alltaxon_path = os.path.join(indir, 'all_taxon.taxonomy')\n",
    "    unrefined_path = os.path.join(indir, 'unrefined.taxonomy')\n",
    "    cmd = f\"python build_taxonomic_tree.py {alltaxon_path} {species_path} {mapping_path} {unrefined_path}\"\n",
    "    p = Popen(cmd, stdout=PIPE, stderr=STDOUT, text=True, shell=True)\n",
    "    for line in p.stdout:\n",
    "        print(line.replace('\\n', ''))\n",
    "    \n",
    "    # rename the taxonomy tree with sequence names from this gene\n",
    "    # with script build_unrefined_tree.pl\n",
    "    taxonomy_path = os.path.join(indir, 'unrefined.taxonomy.renamed')\n",
    "    cmd = f\"perl build_unrefined_tree.pl {mapping_path} {unrefined_path} {taxonomy_path}\"\n",
    "    os.system(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f44bd96b",
   "metadata": {},
   "source": [
    "# (for tipp3) Get RAxML-ng maximum likelihood tree and model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "53ee7136",
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in genes:\n",
    "    indir = os.path.join(workdir, gene)\n",
    "    \n",
    "    # masked the alignment first with >= 95% gaps\n",
    "    aln_path = os.path.join(indir, 'est.aln.nuc.fasta.gz') \n",
    "    masked_path = os.path.join(indir, 'est.aln.nuc.masked95.fasta')\n",
    "    thres = 0.95\n",
    "    cmd = f\"python mask_alignment.py {aln_path} {masked_path} {thres}\"\n",
    "    os.system(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6b3a7451",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for gene in genes:\n",
    "    # run raxml-ng\n",
    "    indir = os.path.join(workdir, gene)\n",
    "    masked_path = os.path.join(indir, 'est.aln.nuc.masked95.fasta')\n",
    "    check_path = os.path.join(indir, 'est.raxml.bestTree')\n",
    "    if not os.path.exists(check_path):\n",
    "        # resolve polytomies in the taxonomy tree and remove internal node labels to run RAxML-ng with\n",
    "        taxonomy_path = os.path.join(indir, 'unrefined.taxonomy.renamed')\n",
    "        starting_tree_path = os.path.join(indir, 'raxml.starting.tre')\n",
    "        cmd = f\"python resolve_polytomies.py {taxonomy_path} {starting_tree_path}\"\n",
    "        os.system(cmd)\n",
    "\n",
    "        # Run RAxML-ng with the starting tree and the original taxonomy tree as the constraint\n",
    "        cmd = ['raxml-ng', '--msa', masked_path, '--model GTR+G',\n",
    "               '--tree', starting_tree_path, '--tree-constraint', taxonomy_path,\n",
    "               '--brlen scaled',\n",
    "               '--force perf_threads', '--redo',\n",
    "               f'--prefix {indir}/est',\n",
    "               '--threads', str(num_threads)]\n",
    "\n",
    "        p = Popen(' '.join(cmd), stdout=PIPE, stderr=STDOUT, text=True, shell=True)\n",
    "        for line in p.stdout:\n",
    "            if not line.startswith('WARNING'):\n",
    "                print(line.replace('\\n', ''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a6096dee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>>> RplA_COG0081\n",
      "reseeding example_data/RplA_COG0081/est.raxml.bestTree\n",
      "relabeling example_data/RplA_COG0081/est.raxml.bestTree\n",
      "\tmapped 637\n",
      "\tmissing 0\n",
      "sanity check correctness True\n",
      "\n",
      ">>>> RplB_COG0090\n",
      "reseeding example_data/RplB_COG0090/est.raxml.bestTree (only one superkingdom group)\n",
      "relabeling example_data/RplB_COG0090/est.raxml.bestTree\n",
      "\tmapped 653\n",
      "\tmissing 0\n",
      "sanity check correctness True\n",
      "\n",
      ">>>> RplD_COG0088\n",
      "reseeding example_data/RplD_COG0088/est.raxml.bestTree (only one superkingdom group)\n",
      "relabeling example_data/RplD_COG0088/est.raxml.bestTree\n",
      "\tmapped 635\n",
      "\tmissing 0\n",
      "sanity check correctness True\n",
      "\n",
      ">>>> RpsC_COG0092\n",
      "reseeding example_data/RpsC_COG0092/est.raxml.bestTree\n",
      "relabeling example_data/RpsC_COG0092/est.raxml.bestTree\n",
      "\tmapped 641\n",
      "\tmissing 0\n",
      "sanity check correctness True\n",
      "\n",
      ">>>> RpsM_COG0099\n",
      "reseeding example_data/RpsM_COG0099/est.raxml.bestTree (only one superkingdom group)\n",
      "relabeling example_data/RpsM_COG0099/est.raxml.bestTree\n",
      "\tmapped 664\n",
      "\tmissing 0\n",
      "sanity check correctness True\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# add back the taxonomy information to the raxml output tree, and reroot the tree at taxid 131567\n",
    "for gene in genes:\n",
    "    print('>>>>', gene)\n",
    "    indir = os.path.join(workdir, gene)\n",
    "    \n",
    "    # reroot the raxml-ng bestTree to 131567 (cellular organism) and relabel the tree\n",
    "    root = 131567\n",
    "    best_tree_path = os.path.join(indir, 'est.raxml.bestTree')\n",
    "    taxonomy_path = os.path.join(indir, 'unrefined.taxonomy.renamed')\n",
    "    rooted_tree_path = os.path.join(indir, 'est.raxml.bestTree.rooted')\n",
    "    cmd = f\"python relabel_tree.py {taxonomy_path} {best_tree_path} {rooted_tree_path} {root}\"\n",
    "    ret = os.popen(cmd).read()\n",
    "    print(ret)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f69a0018",
   "metadata": {},
   "source": [
    "# (for tipp3-accurate) Get FastTree-2 maximum likelihood tree and log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ff0e98bd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>>> RplA_COG0081\n",
      "FastTree Version 2.1.11 SSE3, OpenMP (12 threads)\n",
      "Alignment: standard input\n",
      "Nucleotide distances: Jukes-Cantor Joins: balanced Support: none\n",
      "Start at tree from example_data/RplA_COG0081/est.raxml.bestTree.rooted (no NNI) (no SPR)\n",
      "ML Model: Generalized Time-Reversible, CAT approximation with 20 rate categories\n",
      "Total branch-length 174.190 after 0.13 sec\n",
      "      0.14 seconds: ML Lengths 1 of 1897 splits\n",
      "      0.26 seconds: ML Lengths 401 of 1897 splits\n",
      "      0.38 seconds: ML Lengths 801 of 1897 splits\n",
      "      0.50 seconds: ML Lengths 1201 of 1897 splits\n",
      "      0.63 seconds: ML Lengths 1601 of 1897 splits\n",
      "1 rounds ML lengths: LogLk = -540801.045 Max-change 0.2614 Time 0.74\n",
      "      0.73 seconds: Optimizing GTR model, step 1 of 12\n",
      "      1.28 seconds: Optimizing GTR model, step 2 of 12\n",
      "      1.65 seconds: Optimizing GTR model, step 3 of 12\n",
      "      2.03 seconds: Optimizing GTR model, step 4 of 12\n",
      "      2.49 seconds: Optimizing GTR model, step 5 of 12\n",
      "      3.07 seconds: Optimizing GTR model, step 6 of 12\n",
      "      3.69 seconds: Optimizing GTR model, step 7 of 12\n",
      "      4.07 seconds: Optimizing GTR model, step 8 of 12\n",
      "      4.40 seconds: Optimizing GTR model, step 9 of 12\n",
      "      4.81 seconds: Optimizing GTR model, step 10 of 12\n",
      "      5.19 seconds: Optimizing GTR model, step 11 of 12\n",
      "      5.60 seconds: Optimizing GTR model, step 12 of 12\n",
      "GTR Frequencies: 0.2634 0.2490 0.2750 0.2125\n",
      "GTR rates(ac ag at cg ct gt) 1.3099 2.2788 2.4255 1.9304 4.4465 1.0000\n",
      "      5.97 seconds: ML Lengths 1 of 1897 splits\n",
      "      6.09 seconds: ML Lengths 301 of 1897 splits\n",
      "      6.21 seconds: ML Lengths 601 of 1897 splits\n",
      "      6.34 seconds: ML Lengths 901 of 1897 splits\n",
      "      6.47 seconds: ML Lengths 1201 of 1897 splits\n",
      "      6.60 seconds: ML Lengths 1501 of 1897 splits\n",
      "      6.72 seconds: ML Lengths 1801 of 1897 splits\n",
      "      6.85 seconds: Site likelihoods with rate category 2 of 20\n",
      "      6.98 seconds: Site likelihoods with rate category 5 of 20\n",
      "      7.11 seconds: Site likelihoods with rate category 8 of 20\n",
      "      7.23 seconds: Site likelihoods with rate category 11 of 20\n",
      "      7.36 seconds: Site likelihoods with rate category 14 of 20\n",
      "      7.49 seconds: Site likelihoods with rate category 17 of 20\n",
      "      7.62 seconds: Site likelihoods with rate category 20 of 20\n",
      "Switched to using 20 rate categories (CAT approximation)\n",
      "Rate categories were divided by 1.030 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "      7.73 seconds: ML Lengths 101 of 1897 splits\n",
      "      7.86 seconds: ML Lengths 401 of 1897 splits\n",
      "      7.99 seconds: ML Lengths 701 of 1897 splits\n",
      "      8.13 seconds: ML Lengths 1001 of 1897 splits\n",
      "      8.27 seconds: ML Lengths 1301 of 1897 splits\n",
      "      8.40 seconds: ML Lengths 1601 of 1897 splits\n",
      "2 rounds ML lengths: LogLk = -470652.935 Max-change 1.6940 Time 8.55\n",
      "      8.54 seconds: ML Lengths 1 of 1897 splits\n",
      "      8.67 seconds: ML Lengths 301 of 1897 splits\n",
      "      8.80 seconds: ML Lengths 601 of 1897 splits\n",
      "      8.92 seconds: ML Lengths 901 of 1897 splits\n",
      "      9.06 seconds: ML Lengths 1201 of 1897 splits\n",
      "      9.18 seconds: ML Lengths 1501 of 1897 splits\n",
      "      9.31 seconds: ML Lengths 1801 of 1897 splits\n",
      "3 rounds ML lengths: LogLk = -470635.663 Max-change 0.0902 Time 9.37\n",
      "      9.45 seconds: ML Lengths 201 of 1897 splits\n",
      "      9.57 seconds: ML Lengths 501 of 1897 splits\n",
      "      9.70 seconds: ML Lengths 801 of 1897 splits\n",
      "      9.82 seconds: ML Lengths 1101 of 1897 splits\n",
      "      9.95 seconds: ML Lengths 1401 of 1897 splits\n",
      "     10.08 seconds: ML Lengths 1701 of 1897 splits\n",
      "4 rounds ML lengths: LogLk = -470634.271 Max-change 0.0165 Time 10.18\n",
      "     10.21 seconds: ML Lengths 101 of 1897 splits\n",
      "     10.34 seconds: ML Lengths 401 of 1897 splits\n",
      "     10.46 seconds: ML Lengths 701 of 1897 splits\n",
      "     10.59 seconds: ML Lengths 1001 of 1897 splits\n",
      "     10.71 seconds: ML Lengths 1301 of 1897 splits\n",
      "     10.84 seconds: ML Lengths 1601 of 1897 splits\n",
      "5 rounds ML lengths: LogLk = -470634.109 Max-change 0.0047 Time 10.98\n",
      "     10.97 seconds: ML Lengths 1 of 1897 splits\n",
      "     11.10 seconds: ML Lengths 301 of 1897 splits\n",
      "     11.22 seconds: ML Lengths 601 of 1897 splits\n",
      "     11.35 seconds: ML Lengths 901 of 1897 splits\n",
      "     11.47 seconds: ML Lengths 1201 of 1897 splits\n",
      "     11.60 seconds: ML Lengths 1501 of 1897 splits\n",
      "     11.73 seconds: ML Lengths 1801 of 1897 splits\n",
      "6 rounds ML lengths: LogLk = -470634.079 Max-change 0.0018 Time 11.78\n",
      "     11.86 seconds: ML Lengths 201 of 1897 splits\n",
      "     11.98 seconds: ML Lengths 501 of 1897 splits\n",
      "     12.11 seconds: ML Lengths 801 of 1897 splits\n",
      "     12.23 seconds: ML Lengths 1101 of 1897 splits\n",
      "     12.36 seconds: ML Lengths 1401 of 1897 splits\n",
      "     12.48 seconds: ML Lengths 1701 of 1897 splits\n",
      "7 rounds ML lengths: LogLk = -470634.068 Max-change 0.0010 Time 12.58\n",
      "     12.62 seconds: ML Lengths 101 of 1897 splits\n",
      "     12.74 seconds: ML Lengths 401 of 1897 splits\n",
      "     12.87 seconds: ML Lengths 701 of 1897 splits\n",
      "     12.99 seconds: ML Lengths 1001 of 1897 splits\n",
      "     13.12 seconds: ML Lengths 1301 of 1897 splits\n",
      "     13.25 seconds: ML Lengths 1601 of 1897 splits\n",
      "8 rounds ML lengths: LogLk = -470634.066 Max-change 0.0005 (converged) Time 13.38\n",
      "     13.42 seconds: Site likelihoods with rate category 1 of 20\n",
      "     13.56 seconds: Site likelihoods with rate category 4 of 20\n",
      "     13.69 seconds: Site likelihoods with rate category 7 of 20\n",
      "     13.83 seconds: Site likelihoods with rate category 10 of 20\n",
      "     13.96 seconds: Site likelihoods with rate category 13 of 20\n",
      "     14.09 seconds: Site likelihoods with rate category 16 of 20\n",
      "     14.23 seconds: Site likelihoods with rate category 19 of 20\n",
      "Gamma(20) LogLk = -472346.800 alpha = 0.633 rescaling lengths by 1.213\n",
      "Total time: 14.32 seconds Unique: 1899/2749 Bad splits: 0/0\n",
      ">>>> RplB_COG0090\n",
      "FastTree Version 2.1.11 SSE3, OpenMP (12 threads)\n",
      "Alignment: standard input\n",
      "Nucleotide distances: Jukes-Cantor Joins: balanced Support: none\n",
      "Start at tree from example_data/RplB_COG0090/est.raxml.bestTree.rooted (no NNI) (no SPR)\n",
      "ML Model: Generalized Time-Reversible, CAT approximation with 20 rate categories\n",
      "Ignored unknown character X (seen 1 times)\n",
      "Total branch-length 162.414 after 0.16 sec\n",
      "      0.17 seconds: ML Lengths 1 of 1878 splits\n",
      "      0.28 seconds: ML Lengths 301 of 1878 splits\n",
      "      0.38 seconds: ML Lengths 601 of 1878 splits\n",
      "      0.50 seconds: ML Lengths 901 of 1878 splits\n",
      "      0.62 seconds: ML Lengths 1201 of 1878 splits\n",
      "      0.73 seconds: ML Lengths 1501 of 1878 splits\n",
      "      0.85 seconds: ML Lengths 1801 of 1878 splits\n",
      "1 rounds ML lengths: LogLk = -611118.661 Max-change 0.2415 Time 0.90\n",
      "      1.54 seconds: Optimizing GTR model, step 2 of 12\n",
      "      2.03 seconds: Optimizing GTR model, step 3 of 12\n",
      "      2.52 seconds: Optimizing GTR model, step 4 of 12\n",
      "      3.21 seconds: Optimizing GTR model, step 5 of 12\n",
      "      3.90 seconds: Optimizing GTR model, step 6 of 12\n",
      "      4.73 seconds: Optimizing GTR model, step 7 of 12\n",
      "      5.18 seconds: Optimizing GTR model, step 8 of 12\n",
      "      5.57 seconds: Optimizing GTR model, step 9 of 12\n",
      "      6.06 seconds: Optimizing GTR model, step 10 of 12\n",
      "      6.50 seconds: Optimizing GTR model, step 11 of 12\n",
      "      6.99 seconds: Optimizing GTR model, step 12 of 12\n",
      "GTR Frequencies: 0.2541 0.2654 0.2753 0.2053\n",
      "GTR rates(ac ag at cg ct gt) 1.2223 2.1278 2.5695 1.5886 4.7636 1.0000\n",
      "      7.47 seconds: ML Lengths 1 of 1878 splits\n",
      "      7.58 seconds: ML Lengths 201 of 1878 splits\n",
      "      7.72 seconds: ML Lengths 501 of 1878 splits\n",
      "      7.83 seconds: ML Lengths 701 of 1878 splits\n",
      "      7.93 seconds: ML Lengths 901 of 1878 splits\n",
      "      8.03 seconds: ML Lengths 1101 of 1878 splits\n",
      "      8.14 seconds: ML Lengths 1301 of 1878 splits\n",
      "      8.24 seconds: ML Lengths 1501 of 1878 splits\n",
      "      8.35 seconds: ML Lengths 1701 of 1878 splits\n",
      "      8.49 seconds: Site likelihoods with rate category 1 of 20\n",
      "      8.59 seconds: Site likelihoods with rate category 3 of 20\n",
      "      8.69 seconds: Site likelihoods with rate category 5 of 20\n",
      "      8.79 seconds: Site likelihoods with rate category 7 of 20\n",
      "      8.89 seconds: Site likelihoods with rate category 9 of 20\n",
      "      8.99 seconds: Site likelihoods with rate category 11 of 20\n",
      "      9.09 seconds: Site likelihoods with rate category 13 of 20\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      9.19 seconds: Site likelihoods with rate category 15 of 20\n",
      "      9.30 seconds: Site likelihoods with rate category 17 of 20\n",
      "      9.40 seconds: Site likelihoods with rate category 19 of 20\n",
      "Switched to using 20 rate categories (CAT approximation)\n",
      "Rate categories were divided by 1.002 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "      9.53 seconds: ML Lengths 1 of 1878 splits\n",
      "      9.64 seconds: ML Lengths 201 of 1878 splits\n",
      "      9.74 seconds: ML Lengths 401 of 1878 splits\n",
      "      9.85 seconds: ML Lengths 601 of 1878 splits\n",
      "      9.96 seconds: ML Lengths 801 of 1878 splits\n",
      "     10.07 seconds: ML Lengths 1001 of 1878 splits\n",
      "     10.18 seconds: ML Lengths 1201 of 1878 splits\n",
      "     10.28 seconds: ML Lengths 1401 of 1878 splits\n",
      "     10.40 seconds: ML Lengths 1601 of 1878 splits\n",
      "     10.50 seconds: ML Lengths 1801 of 1878 splits\n",
      "2 rounds ML lengths: LogLk = -530780.097 Max-change 0.4837 Time 10.56\n",
      "     10.61 seconds: ML Lengths 101 of 1878 splits\n",
      "     10.71 seconds: ML Lengths 301 of 1878 splits\n",
      "     10.86 seconds: ML Lengths 601 of 1878 splits\n",
      "     10.97 seconds: ML Lengths 801 of 1878 splits\n",
      "     11.07 seconds: ML Lengths 1001 of 1878 splits\n",
      "     11.17 seconds: ML Lengths 1201 of 1878 splits\n",
      "     11.28 seconds: ML Lengths 1401 of 1878 splits\n",
      "     11.38 seconds: ML Lengths 1601 of 1878 splits\n",
      "     11.49 seconds: ML Lengths 1801 of 1878 splits\n",
      "3 rounds ML lengths: LogLk = -530768.366 Max-change 0.1238 Time 11.54\n",
      "     11.59 seconds: ML Lengths 101 of 1878 splits\n",
      "     11.69 seconds: ML Lengths 301 of 1878 splits\n",
      "     11.84 seconds: ML Lengths 601 of 1878 splits\n",
      "     11.94 seconds: ML Lengths 801 of 1878 splits\n",
      "     12.04 seconds: ML Lengths 1001 of 1878 splits\n",
      "     12.15 seconds: ML Lengths 1201 of 1878 splits\n",
      "     12.25 seconds: ML Lengths 1401 of 1878 splits\n",
      "     12.35 seconds: ML Lengths 1601 of 1878 splits\n",
      "     12.45 seconds: ML Lengths 1801 of 1878 splits\n",
      "4 rounds ML lengths: LogLk = -530767.724 Max-change 0.0350 Time 12.51\n",
      "     12.56 seconds: ML Lengths 101 of 1878 splits\n",
      "     12.71 seconds: ML Lengths 401 of 1878 splits\n",
      "     12.86 seconds: ML Lengths 701 of 1878 splits\n",
      "     12.96 seconds: ML Lengths 901 of 1878 splits\n",
      "     13.06 seconds: ML Lengths 1101 of 1878 splits\n",
      "     13.16 seconds: ML Lengths 1301 of 1878 splits\n",
      "     13.27 seconds: ML Lengths 1501 of 1878 splits\n",
      "     13.37 seconds: ML Lengths 1701 of 1878 splits\n",
      "5 rounds ML lengths: LogLk = -530767.622 Max-change 0.0096 Time 13.47\n",
      "     13.47 seconds: ML Lengths 1 of 1878 splits\n",
      "     13.57 seconds: ML Lengths 201 of 1878 splits\n",
      "     13.72 seconds: ML Lengths 501 of 1878 splits\n",
      "     13.82 seconds: ML Lengths 701 of 1878 splits\n",
      "     13.92 seconds: ML Lengths 901 of 1878 splits\n",
      "     14.02 seconds: ML Lengths 1101 of 1878 splits\n",
      "     14.13 seconds: ML Lengths 1301 of 1878 splits\n",
      "     14.23 seconds: ML Lengths 1501 of 1878 splits\n",
      "     14.33 seconds: ML Lengths 1701 of 1878 splits\n",
      "6 rounds ML lengths: LogLk = -530767.590 Max-change 0.0035 Time 14.44\n",
      "     14.43 seconds: ML Lengths 1 of 1878 splits\n",
      "     14.53 seconds: ML Lengths 201 of 1878 splits\n",
      "     14.68 seconds: ML Lengths 501 of 1878 splits\n",
      "     14.78 seconds: ML Lengths 701 of 1878 splits\n",
      "     14.88 seconds: ML Lengths 901 of 1878 splits\n",
      "     14.99 seconds: ML Lengths 1101 of 1878 splits\n",
      "     15.09 seconds: ML Lengths 1301 of 1878 splits\n",
      "     15.19 seconds: ML Lengths 1501 of 1878 splits\n",
      "     15.29 seconds: ML Lengths 1701 of 1878 splits\n",
      "7 rounds ML lengths: LogLk = -530767.583 Max-change 0.0010 (converged) Time 15.40\n",
      "     15.45 seconds: Site likelihoods with rate category 1 of 20\n",
      "     15.55 seconds: Site likelihoods with rate category 3 of 20\n",
      "     15.66 seconds: Site likelihoods with rate category 5 of 20\n",
      "     15.76 seconds: Site likelihoods with rate category 7 of 20\n",
      "     15.87 seconds: Site likelihoods with rate category 9 of 20\n",
      "     15.97 seconds: Site likelihoods with rate category 11 of 20\n",
      "     16.08 seconds: Site likelihoods with rate category 13 of 20\n",
      "     16.18 seconds: Site likelihoods with rate category 15 of 20\n",
      "     16.29 seconds: Site likelihoods with rate category 17 of 20\n",
      "     16.39 seconds: Site likelihoods with rate category 19 of 20\n",
      "Gamma(20) LogLk = -532784.257 alpha = 0.621 rescaling lengths by 1.223\n",
      "Total time: 16.50 seconds Unique: 1880/2755 Bad splits: 0/0\n",
      ">>>> RplD_COG0088\n",
      "FastTree Version 2.1.11 SSE3, OpenMP (12 threads)\n",
      "Alignment: standard input\n",
      "Nucleotide distances: Jukes-Cantor Joins: balanced Support: none\n",
      "Start at tree from example_data/RplD_COG0088/est.raxml.bestTree.rooted (no NNI) (no SPR)\n",
      "ML Model: Generalized Time-Reversible, CAT approximation with 20 rate categories\n",
      "Ignored unknown character M (seen 1 times)\n",
      "Ignored unknown character Y (seen 2 times)\n",
      "Total branch-length 198.403 after 0.17 sec\n",
      "      0.19 seconds: ML Lengths 1 of 1788 splits\n",
      "      0.30 seconds: ML Lengths 201 of 1788 splits\n",
      "      0.42 seconds: ML Lengths 401 of 1788 splits\n",
      "      0.54 seconds: ML Lengths 601 of 1788 splits\n",
      "      0.65 seconds: ML Lengths 801 of 1788 splits\n",
      "      0.76 seconds: ML Lengths 1001 of 1788 splits\n",
      "      0.87 seconds: ML Lengths 1201 of 1788 splits\n",
      "      0.98 seconds: ML Lengths 1401 of 1788 splits\n",
      "      1.10 seconds: ML Lengths 1601 of 1788 splits\n",
      "1 rounds ML lengths: LogLk = -503530.433 Max-change 0.2387 Time 1.23\n",
      "      1.23 seconds: Optimizing GTR model, step 1 of 12\n",
      "      1.79 seconds: Optimizing GTR model, step 2 of 12\n",
      "      2.21 seconds: Optimizing GTR model, step 3 of 12\n",
      "      2.64 seconds: Optimizing GTR model, step 4 of 12\n",
      "      3.07 seconds: Optimizing GTR model, step 5 of 12\n",
      "      3.66 seconds: Optimizing GTR model, step 6 of 12\n",
      "      4.30 seconds: Optimizing GTR model, step 7 of 12\n",
      "      4.64 seconds: Optimizing GTR model, step 8 of 12\n",
      "      5.02 seconds: Optimizing GTR model, step 9 of 12\n",
      "      5.41 seconds: Optimizing GTR model, step 10 of 12\n",
      "      5.79 seconds: Optimizing GTR model, step 11 of 12\n",
      "      6.22 seconds: Optimizing GTR model, step 12 of 12\n",
      "GTR Frequencies: 0.2613 0.2515 0.2694 0.2179\n",
      "GTR rates(ac ag at cg ct gt) 1.3958 2.3245 2.0301 1.9750 3.8419 1.0000\n",
      "      6.59 seconds: ML Lengths 1 of 1788 splits\n",
      "      6.74 seconds: ML Lengths 201 of 1788 splits\n",
      "      6.89 seconds: ML Lengths 401 of 1788 splits\n",
      "      7.04 seconds: ML Lengths 601 of 1788 splits\n",
      "      7.19 seconds: ML Lengths 801 of 1788 splits\n",
      "      7.34 seconds: ML Lengths 1001 of 1788 splits\n",
      "      7.49 seconds: ML Lengths 1201 of 1788 splits\n",
      "      7.64 seconds: ML Lengths 1401 of 1788 splits\n",
      "      7.79 seconds: ML Lengths 1601 of 1788 splits\n",
      "      7.98 seconds: Site likelihoods with rate category 1 of 20\n",
      "      8.11 seconds: Site likelihoods with rate category 4 of 20\n",
      "      8.24 seconds: Site likelihoods with rate category 7 of 20\n",
      "      8.38 seconds: Site likelihoods with rate category 10 of 20\n",
      "      8.51 seconds: Site likelihoods with rate category 13 of 20\n",
      "      8.64 seconds: Site likelihoods with rate category 16 of 20\n",
      "      8.78 seconds: Site likelihoods with rate category 19 of 20\n",
      "Switched to using 20 rate categories (CAT approximation)\n",
      "Rate categories were divided by 0.891 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "      8.89 seconds: ML Lengths 1 of 1788 splits\n",
      "      9.05 seconds: ML Lengths 201 of 1788 splits\n",
      "      9.21 seconds: ML Lengths 401 of 1788 splits\n",
      "      9.37 seconds: ML Lengths 601 of 1788 splits\n",
      "      9.53 seconds: ML Lengths 801 of 1788 splits\n",
      "      9.68 seconds: ML Lengths 1001 of 1788 splits\n",
      "      9.84 seconds: ML Lengths 1201 of 1788 splits\n",
      "      9.99 seconds: ML Lengths 1401 of 1788 splits\n",
      "     10.15 seconds: ML Lengths 1601 of 1788 splits\n",
      "2 rounds ML lengths: LogLk = -455771.219 Max-change 0.5912 Time 10.32\n",
      "     10.32 seconds: ML Lengths 1 of 1788 splits\n",
      "     10.47 seconds: ML Lengths 201 of 1788 splits\n",
      "     10.62 seconds: ML Lengths 401 of 1788 splits\n",
      "     10.77 seconds: ML Lengths 601 of 1788 splits\n",
      "     10.92 seconds: ML Lengths 801 of 1788 splits\n",
      "     11.07 seconds: ML Lengths 1001 of 1788 splits\n",
      "     11.22 seconds: ML Lengths 1201 of 1788 splits\n",
      "     11.36 seconds: ML Lengths 1401 of 1788 splits\n",
      "     11.51 seconds: ML Lengths 1601 of 1788 splits\n",
      "3 rounds ML lengths: LogLk = -455764.273 Max-change 0.0676 Time 11.67\n",
      "     11.67 seconds: ML Lengths 1 of 1788 splits\n",
      "     11.82 seconds: ML Lengths 201 of 1788 splits\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     11.97 seconds: ML Lengths 401 of 1788 splits\n",
      "     12.11 seconds: ML Lengths 601 of 1788 splits\n",
      "     12.26 seconds: ML Lengths 801 of 1788 splits\n",
      "     12.40 seconds: ML Lengths 1001 of 1788 splits\n",
      "     12.55 seconds: ML Lengths 1201 of 1788 splits\n",
      "     12.69 seconds: ML Lengths 1401 of 1788 splits\n",
      "     12.84 seconds: ML Lengths 1601 of 1788 splits\n",
      "4 rounds ML lengths: LogLk = -455763.841 Max-change 0.0191 Time 13.00\n",
      "     12.99 seconds: ML Lengths 1 of 1788 splits\n",
      "     13.14 seconds: ML Lengths 201 of 1788 splits\n",
      "     13.29 seconds: ML Lengths 401 of 1788 splits\n",
      "     13.44 seconds: ML Lengths 601 of 1788 splits\n",
      "     13.58 seconds: ML Lengths 801 of 1788 splits\n",
      "     13.73 seconds: ML Lengths 1001 of 1788 splits\n",
      "     13.87 seconds: ML Lengths 1201 of 1788 splits\n",
      "     14.02 seconds: ML Lengths 1401 of 1788 splits\n",
      "     14.16 seconds: ML Lengths 1601 of 1788 splits\n",
      "5 rounds ML lengths: LogLk = -455763.699 Max-change 0.0088 Time 14.32\n",
      "     14.32 seconds: ML Lengths 1 of 1788 splits\n",
      "     14.47 seconds: ML Lengths 201 of 1788 splits\n",
      "     14.61 seconds: ML Lengths 401 of 1788 splits\n",
      "     14.76 seconds: ML Lengths 601 of 1788 splits\n",
      "     14.91 seconds: ML Lengths 801 of 1788 splits\n",
      "     15.05 seconds: ML Lengths 1001 of 1788 splits\n",
      "     15.19 seconds: ML Lengths 1201 of 1788 splits\n",
      "     15.34 seconds: ML Lengths 1401 of 1788 splits\n",
      "     15.49 seconds: ML Lengths 1601 of 1788 splits\n",
      "6 rounds ML lengths: LogLk = -455763.607 Max-change 0.0044 Time 15.64\n",
      "     15.64 seconds: ML Lengths 1 of 1788 splits\n",
      "     15.79 seconds: ML Lengths 201 of 1788 splits\n",
      "     15.93 seconds: ML Lengths 401 of 1788 splits\n",
      "     16.08 seconds: ML Lengths 601 of 1788 splits\n",
      "     16.23 seconds: ML Lengths 801 of 1788 splits\n",
      "     16.37 seconds: ML Lengths 1001 of 1788 splits\n",
      "     16.52 seconds: ML Lengths 1201 of 1788 splits\n",
      "     16.66 seconds: ML Lengths 1401 of 1788 splits\n",
      "     16.81 seconds: ML Lengths 1601 of 1788 splits\n",
      "7 rounds ML lengths: LogLk = -455763.551 Max-change 0.0030 Time 16.96\n",
      "     16.96 seconds: ML Lengths 1 of 1788 splits\n",
      "     17.11 seconds: ML Lengths 201 of 1788 splits\n",
      "     17.25 seconds: ML Lengths 401 of 1788 splits\n",
      "     17.40 seconds: ML Lengths 601 of 1788 splits\n",
      "     17.55 seconds: ML Lengths 801 of 1788 splits\n",
      "     17.69 seconds: ML Lengths 1001 of 1788 splits\n",
      "     17.84 seconds: ML Lengths 1201 of 1788 splits\n",
      "     17.98 seconds: ML Lengths 1401 of 1788 splits\n",
      "     18.13 seconds: ML Lengths 1601 of 1788 splits\n",
      "8 rounds ML lengths: LogLk = -455763.521 Max-change 0.0039 Time 18.28\n",
      "     18.28 seconds: ML Lengths 1 of 1788 splits\n",
      "     18.43 seconds: ML Lengths 201 of 1788 splits\n",
      "     18.58 seconds: ML Lengths 401 of 1788 splits\n",
      "     18.72 seconds: ML Lengths 601 of 1788 splits\n",
      "     18.87 seconds: ML Lengths 801 of 1788 splits\n",
      "     19.01 seconds: ML Lengths 1001 of 1788 splits\n",
      "     19.16 seconds: ML Lengths 1201 of 1788 splits\n",
      "     19.30 seconds: ML Lengths 1401 of 1788 splits\n",
      "     19.45 seconds: ML Lengths 1601 of 1788 splits\n",
      "9 rounds ML lengths: LogLk = -455763.495 Max-change 0.0025 Time 19.60\n",
      "     19.60 seconds: ML Lengths 1 of 1788 splits\n",
      "     19.75 seconds: ML Lengths 201 of 1788 splits\n",
      "     19.90 seconds: ML Lengths 401 of 1788 splits\n",
      "     20.04 seconds: ML Lengths 601 of 1788 splits\n",
      "     20.19 seconds: ML Lengths 801 of 1788 splits\n",
      "     20.33 seconds: ML Lengths 1001 of 1788 splits\n",
      "     20.48 seconds: ML Lengths 1201 of 1788 splits\n",
      "     20.62 seconds: ML Lengths 1401 of 1788 splits\n",
      "     20.77 seconds: ML Lengths 1601 of 1788 splits\n",
      "10 rounds ML lengths: LogLk = -455763.474 Max-change 0.0024 Time 20.92\n",
      "     20.92 seconds: ML Lengths 1 of 1788 splits\n",
      "     21.07 seconds: ML Lengths 201 of 1788 splits\n",
      "     21.22 seconds: ML Lengths 401 of 1788 splits\n",
      "     21.36 seconds: ML Lengths 601 of 1788 splits\n",
      "     21.51 seconds: ML Lengths 801 of 1788 splits\n",
      "     21.65 seconds: ML Lengths 1001 of 1788 splits\n",
      "     21.80 seconds: ML Lengths 1201 of 1788 splits\n",
      "     21.94 seconds: ML Lengths 1401 of 1788 splits\n",
      "     22.09 seconds: ML Lengths 1601 of 1788 splits\n",
      "11 rounds ML lengths: LogLk = -455763.455 Max-change 0.0022 Time 22.24\n",
      "     22.28 seconds: Site likelihoods with rate category 1 of 20\n",
      "     22.42 seconds: Site likelihoods with rate category 4 of 20\n",
      "     22.56 seconds: Site likelihoods with rate category 7 of 20\n",
      "     22.70 seconds: Site likelihoods with rate category 10 of 20\n",
      "     22.84 seconds: Site likelihoods with rate category 13 of 20\n",
      "     22.98 seconds: Site likelihoods with rate category 16 of 20\n",
      "     23.12 seconds: Site likelihoods with rate category 19 of 20\n",
      "Gamma(20) LogLk = -457381.199 alpha = 1.093 rescaling lengths by 1.618\n",
      "Total time: 23.22 seconds Unique: 1790/2731 Bad splits: 0/0\n",
      ">>>> RpsC_COG0092\n",
      "FastTree Version 2.1.11 SSE3, OpenMP (12 threads)\n",
      "Alignment: standard input\n",
      "Nucleotide distances: Jukes-Cantor Joins: balanced Support: none\n",
      "Start at tree from example_data/RpsC_COG0092/est.raxml.bestTree.rooted (no NNI) (no SPR)\n",
      "ML Model: Generalized Time-Reversible, CAT approximation with 20 rate categories\n",
      "Ignored unknown character R (seen 1 times)\n",
      "Total branch-length 157.953 after 0.39 sec\n",
      "      0.42 seconds: ML Lengths 1 of 1811 splits\n",
      "      0.57 seconds: ML Lengths 101 of 1811 splits\n",
      "      0.73 seconds: ML Lengths 201 of 1811 splits\n",
      "      0.87 seconds: ML Lengths 301 of 1811 splits\n",
      "      1.02 seconds: ML Lengths 401 of 1811 splits\n",
      "      1.18 seconds: ML Lengths 501 of 1811 splits\n",
      "      1.33 seconds: ML Lengths 601 of 1811 splits\n",
      "      1.49 seconds: ML Lengths 701 of 1811 splits\n",
      "      1.64 seconds: ML Lengths 801 of 1811 splits\n",
      "      1.80 seconds: ML Lengths 901 of 1811 splits\n",
      "      1.94 seconds: ML Lengths 1001 of 1811 splits\n",
      "      2.07 seconds: ML Lengths 1101 of 1811 splits\n",
      "      2.22 seconds: ML Lengths 1201 of 1811 splits\n",
      "      2.38 seconds: ML Lengths 1301 of 1811 splits\n",
      "      2.54 seconds: ML Lengths 1401 of 1811 splits\n",
      "      2.69 seconds: ML Lengths 1501 of 1811 splits\n",
      "      2.84 seconds: ML Lengths 1601 of 1811 splits\n",
      "      3.00 seconds: ML Lengths 1701 of 1811 splits\n",
      "      3.16 seconds: ML Lengths 1801 of 1811 splits\n",
      "1 rounds ML lengths: LogLk = -512675.126 Max-change 0.2292 Time 3.25\n",
      "      4.14 seconds: Optimizing GTR model, step 2 of 12\n",
      "      4.76 seconds: Optimizing GTR model, step 3 of 12\n",
      "      5.44 seconds: Optimizing GTR model, step 4 of 12\n",
      "      6.26 seconds: Optimizing GTR model, step 5 of 12\n",
      "      7.15 seconds: Optimizing GTR model, step 6 of 12\n",
      "      8.31 seconds: Optimizing GTR model, step 7 of 12\n",
      "      8.93 seconds: Optimizing GTR model, step 8 of 12\n",
      "      9.54 seconds: Optimizing GTR model, step 9 of 12\n",
      "     10.22 seconds: Optimizing GTR model, step 10 of 12\n",
      "     10.84 seconds: Optimizing GTR model, step 11 of 12\n",
      "     11.52 seconds: Optimizing GTR model, step 12 of 12\n",
      "GTR Frequencies: 0.2631 0.2485 0.2822 0.2063\n",
      "GTR rates(ac ag at cg ct gt) 1.5664 2.5228 2.4459 2.0301 5.3809 1.0000\n",
      "     12.19 seconds: ML Lengths 1 of 1811 splits\n",
      "     12.38 seconds: ML Lengths 101 of 1811 splits\n",
      "     12.56 seconds: ML Lengths 201 of 1811 splits\n",
      "     12.75 seconds: ML Lengths 301 of 1811 splits\n",
      "     12.93 seconds: ML Lengths 401 of 1811 splits\n",
      "     13.12 seconds: ML Lengths 501 of 1811 splits\n",
      "     13.31 seconds: ML Lengths 601 of 1811 splits\n",
      "     13.50 seconds: ML Lengths 701 of 1811 splits\n",
      "     13.68 seconds: ML Lengths 801 of 1811 splits\n",
      "     13.87 seconds: ML Lengths 901 of 1811 splits\n",
      "     14.05 seconds: ML Lengths 1001 of 1811 splits\n",
      "     14.23 seconds: ML Lengths 1101 of 1811 splits\n",
      "     14.41 seconds: ML Lengths 1201 of 1811 splits\n",
      "     14.60 seconds: ML Lengths 1301 of 1811 splits\n",
      "     14.79 seconds: ML Lengths 1401 of 1811 splits\n",
      "     14.97 seconds: ML Lengths 1501 of 1811 splits\n",
      "     15.16 seconds: ML Lengths 1601 of 1811 splits\n",
      "     15.35 seconds: ML Lengths 1701 of 1811 splits\n",
      "     15.54 seconds: ML Lengths 1801 of 1811 splits\n",
      "     15.70 seconds: Site likelihoods with rate category 2 of 20\n",
      "     15.85 seconds: Site likelihoods with rate category 4 of 20\n",
      "     15.99 seconds: Site likelihoods with rate category 6 of 20\n",
      "     16.14 seconds: Site likelihoods with rate category 8 of 20\n",
      "     16.28 seconds: Site likelihoods with rate category 10 of 20\n",
      "     16.42 seconds: Site likelihoods with rate category 12 of 20\n",
      "     16.57 seconds: Site likelihoods with rate category 14 of 20\n",
      "     16.71 seconds: Site likelihoods with rate category 16 of 20\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     16.86 seconds: Site likelihoods with rate category 18 of 20\n",
      "     17.00 seconds: Site likelihoods with rate category 20 of 20\n",
      "Switched to using 20 rate categories (CAT approximation)\n",
      "Rate categories were divided by 0.909 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "     17.10 seconds: ML Lengths 1 of 1811 splits\n",
      "     17.30 seconds: ML Lengths 101 of 1811 splits\n",
      "     17.49 seconds: ML Lengths 201 of 1811 splits\n",
      "     17.68 seconds: ML Lengths 301 of 1811 splits\n",
      "     17.87 seconds: ML Lengths 401 of 1811 splits\n",
      "     18.06 seconds: ML Lengths 501 of 1811 splits\n",
      "     18.27 seconds: ML Lengths 601 of 1811 splits\n",
      "     18.46 seconds: ML Lengths 701 of 1811 splits\n",
      "     18.66 seconds: ML Lengths 801 of 1811 splits\n",
      "     18.85 seconds: ML Lengths 901 of 1811 splits\n",
      "     19.04 seconds: ML Lengths 1001 of 1811 splits\n",
      "     19.22 seconds: ML Lengths 1101 of 1811 splits\n",
      "     19.40 seconds: ML Lengths 1201 of 1811 splits\n",
      "     19.60 seconds: ML Lengths 1301 of 1811 splits\n",
      "     19.80 seconds: ML Lengths 1401 of 1811 splits\n",
      "     19.99 seconds: ML Lengths 1501 of 1811 splits\n",
      "     20.18 seconds: ML Lengths 1601 of 1811 splits\n",
      "     20.37 seconds: ML Lengths 1701 of 1811 splits\n",
      "     20.57 seconds: ML Lengths 1801 of 1811 splits\n",
      "2 rounds ML lengths: LogLk = -447814.545 Max-change 1.2087 Time 20.62\n",
      "     20.80 seconds: ML Lengths 101 of 1811 splits\n",
      "     20.99 seconds: ML Lengths 201 of 1811 splits\n",
      "     21.17 seconds: ML Lengths 301 of 1811 splits\n",
      "     21.34 seconds: ML Lengths 401 of 1811 splits\n",
      "     21.53 seconds: ML Lengths 501 of 1811 splits\n",
      "     21.72 seconds: ML Lengths 601 of 1811 splits\n",
      "     21.90 seconds: ML Lengths 701 of 1811 splits\n",
      "     22.08 seconds: ML Lengths 801 of 1811 splits\n",
      "     22.27 seconds: ML Lengths 901 of 1811 splits\n",
      "     22.44 seconds: ML Lengths 1001 of 1811 splits\n",
      "     22.62 seconds: ML Lengths 1101 of 1811 splits\n",
      "     22.80 seconds: ML Lengths 1201 of 1811 splits\n",
      "     22.98 seconds: ML Lengths 1301 of 1811 splits\n",
      "     23.17 seconds: ML Lengths 1401 of 1811 splits\n",
      "     23.35 seconds: ML Lengths 1501 of 1811 splits\n",
      "     23.53 seconds: ML Lengths 1601 of 1811 splits\n",
      "     23.71 seconds: ML Lengths 1701 of 1811 splits\n",
      "     23.90 seconds: ML Lengths 1801 of 1811 splits\n",
      "3 rounds ML lengths: LogLk = -447802.869 Max-change 0.1344 Time 23.95\n",
      "     24.13 seconds: ML Lengths 101 of 1811 splits\n",
      "     24.31 seconds: ML Lengths 201 of 1811 splits\n",
      "     24.48 seconds: ML Lengths 301 of 1811 splits\n",
      "     24.66 seconds: ML Lengths 401 of 1811 splits\n",
      "     24.84 seconds: ML Lengths 501 of 1811 splits\n",
      "     25.03 seconds: ML Lengths 601 of 1811 splits\n",
      "     25.21 seconds: ML Lengths 701 of 1811 splits\n",
      "     25.39 seconds: ML Lengths 801 of 1811 splits\n",
      "     25.57 seconds: ML Lengths 901 of 1811 splits\n",
      "     25.74 seconds: ML Lengths 1001 of 1811 splits\n",
      "     25.91 seconds: ML Lengths 1101 of 1811 splits\n",
      "     26.09 seconds: ML Lengths 1201 of 1811 splits\n",
      "     26.27 seconds: ML Lengths 1301 of 1811 splits\n",
      "     26.45 seconds: ML Lengths 1401 of 1811 splits\n",
      "     26.63 seconds: ML Lengths 1501 of 1811 splits\n",
      "     26.81 seconds: ML Lengths 1601 of 1811 splits\n",
      "     26.99 seconds: ML Lengths 1701 of 1811 splits\n",
      "     27.17 seconds: ML Lengths 1801 of 1811 splits\n",
      "4 rounds ML lengths: LogLk = -447801.817 Max-change 0.0370 Time 27.22\n",
      "     27.39 seconds: ML Lengths 101 of 1811 splits\n",
      "     27.57 seconds: ML Lengths 201 of 1811 splits\n",
      "     27.75 seconds: ML Lengths 301 of 1811 splits\n",
      "     27.92 seconds: ML Lengths 401 of 1811 splits\n",
      "     28.11 seconds: ML Lengths 501 of 1811 splits\n",
      "     28.29 seconds: ML Lengths 601 of 1811 splits\n",
      "     28.47 seconds: ML Lengths 701 of 1811 splits\n",
      "     28.65 seconds: ML Lengths 801 of 1811 splits\n",
      "     28.83 seconds: ML Lengths 901 of 1811 splits\n",
      "     29.00 seconds: ML Lengths 1001 of 1811 splits\n",
      "     29.17 seconds: ML Lengths 1101 of 1811 splits\n",
      "     29.34 seconds: ML Lengths 1201 of 1811 splits\n",
      "     29.52 seconds: ML Lengths 1301 of 1811 splits\n",
      "     29.70 seconds: ML Lengths 1401 of 1811 splits\n",
      "     29.88 seconds: ML Lengths 1501 of 1811 splits\n",
      "     30.06 seconds: ML Lengths 1601 of 1811 splits\n",
      "     30.24 seconds: ML Lengths 1701 of 1811 splits\n",
      "     30.42 seconds: ML Lengths 1801 of 1811 splits\n",
      "5 rounds ML lengths: LogLk = -447801.639 Max-change 0.0116 Time 30.47\n",
      "     30.64 seconds: ML Lengths 101 of 1811 splits\n",
      "     30.82 seconds: ML Lengths 201 of 1811 splits\n",
      "     31.00 seconds: ML Lengths 301 of 1811 splits\n",
      "     31.17 seconds: ML Lengths 401 of 1811 splits\n",
      "     31.35 seconds: ML Lengths 501 of 1811 splits\n",
      "     31.54 seconds: ML Lengths 601 of 1811 splits\n",
      "     31.72 seconds: ML Lengths 701 of 1811 splits\n",
      "     31.90 seconds: ML Lengths 801 of 1811 splits\n",
      "     32.07 seconds: ML Lengths 901 of 1811 splits\n",
      "     32.25 seconds: ML Lengths 1001 of 1811 splits\n",
      "     32.42 seconds: ML Lengths 1101 of 1811 splits\n",
      "     32.59 seconds: ML Lengths 1201 of 1811 splits\n",
      "     32.77 seconds: ML Lengths 1301 of 1811 splits\n",
      "     32.95 seconds: ML Lengths 1401 of 1811 splits\n",
      "     33.13 seconds: ML Lengths 1501 of 1811 splits\n",
      "     33.31 seconds: ML Lengths 1601 of 1811 splits\n",
      "     33.49 seconds: ML Lengths 1701 of 1811 splits\n",
      "     33.67 seconds: ML Lengths 1801 of 1811 splits\n",
      "6 rounds ML lengths: LogLk = -447801.561 Max-change 0.0035 Time 33.71\n",
      "     33.89 seconds: ML Lengths 101 of 1811 splits\n",
      "     34.07 seconds: ML Lengths 201 of 1811 splits\n",
      "     34.24 seconds: ML Lengths 301 of 1811 splits\n",
      "     34.42 seconds: ML Lengths 401 of 1811 splits\n",
      "     34.60 seconds: ML Lengths 501 of 1811 splits\n",
      "     34.78 seconds: ML Lengths 601 of 1811 splits\n",
      "     34.96 seconds: ML Lengths 701 of 1811 splits\n",
      "     35.14 seconds: ML Lengths 801 of 1811 splits\n",
      "     35.32 seconds: ML Lengths 901 of 1811 splits\n",
      "     35.49 seconds: ML Lengths 1001 of 1811 splits\n",
      "     35.66 seconds: ML Lengths 1101 of 1811 splits\n",
      "     35.84 seconds: ML Lengths 1201 of 1811 splits\n",
      "     36.02 seconds: ML Lengths 1301 of 1811 splits\n",
      "     36.20 seconds: ML Lengths 1401 of 1811 splits\n",
      "     36.37 seconds: ML Lengths 1501 of 1811 splits\n",
      "     36.55 seconds: ML Lengths 1601 of 1811 splits\n",
      "     36.73 seconds: ML Lengths 1701 of 1811 splits\n",
      "     36.91 seconds: ML Lengths 1801 of 1811 splits\n",
      "7 rounds ML lengths: LogLk = -447801.545 Max-change 0.0020 Time 36.96\n",
      "     37.13 seconds: ML Lengths 101 of 1811 splits\n",
      "     37.31 seconds: ML Lengths 201 of 1811 splits\n",
      "     37.49 seconds: ML Lengths 301 of 1811 splits\n",
      "     37.66 seconds: ML Lengths 401 of 1811 splits\n",
      "     37.84 seconds: ML Lengths 501 of 1811 splits\n",
      "     38.03 seconds: ML Lengths 601 of 1811 splits\n",
      "     38.21 seconds: ML Lengths 701 of 1811 splits\n",
      "     38.39 seconds: ML Lengths 801 of 1811 splits\n",
      "     38.56 seconds: ML Lengths 901 of 1811 splits\n",
      "     38.74 seconds: ML Lengths 1001 of 1811 splits\n",
      "     38.91 seconds: ML Lengths 1101 of 1811 splits\n",
      "     39.08 seconds: ML Lengths 1201 of 1811 splits\n",
      "     39.26 seconds: ML Lengths 1301 of 1811 splits\n",
      "     39.44 seconds: ML Lengths 1401 of 1811 splits\n",
      "     39.62 seconds: ML Lengths 1501 of 1811 splits\n",
      "     39.80 seconds: ML Lengths 1601 of 1811 splits\n",
      "     39.98 seconds: ML Lengths 1701 of 1811 splits\n",
      "     40.16 seconds: ML Lengths 1801 of 1811 splits\n",
      "8 rounds ML lengths: LogLk = -447801.539 Max-change 0.0013 Time 40.20\n",
      "     40.38 seconds: ML Lengths 101 of 1811 splits\n",
      "     40.56 seconds: ML Lengths 201 of 1811 splits\n",
      "     40.73 seconds: ML Lengths 301 of 1811 splits\n",
      "     40.91 seconds: ML Lengths 401 of 1811 splits\n",
      "     41.09 seconds: ML Lengths 501 of 1811 splits\n",
      "     41.27 seconds: ML Lengths 601 of 1811 splits\n",
      "     41.45 seconds: ML Lengths 701 of 1811 splits\n",
      "     41.63 seconds: ML Lengths 801 of 1811 splits\n",
      "     41.81 seconds: ML Lengths 901 of 1811 splits\n",
      "     41.98 seconds: ML Lengths 1001 of 1811 splits\n",
      "     42.15 seconds: ML Lengths 1101 of 1811 splits\n",
      "     42.33 seconds: ML Lengths 1201 of 1811 splits\n",
      "     42.51 seconds: ML Lengths 1301 of 1811 splits\n",
      "     42.69 seconds: ML Lengths 1401 of 1811 splits\n",
      "     42.87 seconds: ML Lengths 1501 of 1811 splits\n",
      "     43.04 seconds: ML Lengths 1601 of 1811 splits\n",
      "     43.22 seconds: ML Lengths 1701 of 1811 splits\n",
      "     43.41 seconds: ML Lengths 1801 of 1811 splits\n",
      "9 rounds ML lengths: LogLk = -447801.538 Max-change 0.0010 Time 43.45\n",
      "     43.63 seconds: ML Lengths 101 of 1811 splits\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     43.80 seconds: ML Lengths 201 of 1811 splits\n",
      "     43.98 seconds: ML Lengths 301 of 1811 splits\n",
      "     44.16 seconds: ML Lengths 401 of 1811 splits\n",
      "     44.34 seconds: ML Lengths 501 of 1811 splits\n",
      "     44.52 seconds: ML Lengths 601 of 1811 splits\n",
      "     44.70 seconds: ML Lengths 701 of 1811 splits\n",
      "     44.88 seconds: ML Lengths 801 of 1811 splits\n",
      "     45.05 seconds: ML Lengths 901 of 1811 splits\n",
      "     45.23 seconds: ML Lengths 1001 of 1811 splits\n",
      "     45.40 seconds: ML Lengths 1101 of 1811 splits\n",
      "     45.57 seconds: ML Lengths 1201 of 1811 splits\n",
      "     45.75 seconds: ML Lengths 1301 of 1811 splits\n",
      "     45.93 seconds: ML Lengths 1401 of 1811 splits\n",
      "     46.11 seconds: ML Lengths 1501 of 1811 splits\n",
      "     46.29 seconds: ML Lengths 1601 of 1811 splits\n",
      "     46.47 seconds: ML Lengths 1701 of 1811 splits\n",
      "     46.65 seconds: ML Lengths 1801 of 1811 splits\n",
      "10 rounds ML lengths: LogLk = -447801.536 Max-change 0.0009 (converged) Time 46.70\n",
      "     46.77 seconds: Site likelihoods with rate category 1 of 20\n",
      "     46.91 seconds: Site likelihoods with rate category 3 of 20\n",
      "     47.06 seconds: Site likelihoods with rate category 5 of 20\n",
      "     47.21 seconds: Site likelihoods with rate category 7 of 20\n",
      "     47.36 seconds: Site likelihoods with rate category 9 of 20\n",
      "     47.51 seconds: Site likelihoods with rate category 11 of 20\n",
      "     47.65 seconds: Site likelihoods with rate category 13 of 20\n",
      "     47.80 seconds: Site likelihoods with rate category 15 of 20\n",
      "     47.95 seconds: Site likelihoods with rate category 17 of 20\n",
      "     48.10 seconds: Site likelihoods with rate category 19 of 20\n",
      "     48.22 seconds: Optimizing alpha round 1\n",
      "Gamma(20) LogLk = -449927.972 alpha = 0.946 rescaling lengths by 2.367\n",
      "Total time: 48.30 seconds Unique: 1813/2731 Bad splits: 0/0\n",
      ">>>> RpsM_COG0099\n",
      "FastTree Version 2.1.11 SSE3, OpenMP (12 threads)\n",
      "Alignment: standard input\n",
      "Nucleotide distances: Jukes-Cantor Joins: balanced Support: none\n",
      "Start at tree from example_data/RpsM_COG0099/est.raxml.bestTree.rooted (no NNI) (no SPR)\n",
      "ML Model: Generalized Time-Reversible, CAT approximation with 20 rate categories\n",
      "Total branch-length 153.385 after 0.06 sec\n",
      "      0.10 seconds: ML Lengths 201 of 1702 splits\n",
      "      0.21 seconds: ML Lengths 901 of 1702 splits\n",
      "      0.32 seconds: ML Lengths 1601 of 1702 splits\n",
      "1 rounds ML lengths: LogLk = -253069.744 Max-change 0.2135 Time 0.35\n",
      "      0.60 seconds: Optimizing GTR model, step 2 of 12\n",
      "      0.78 seconds: Optimizing GTR model, step 3 of 12\n",
      "      0.98 seconds: Optimizing GTR model, step 4 of 12\n",
      "      1.25 seconds: Optimizing GTR model, step 5 of 12\n",
      "      1.51 seconds: Optimizing GTR model, step 6 of 12\n",
      "      1.80 seconds: Optimizing GTR model, step 7 of 12\n",
      "      1.98 seconds: Optimizing GTR model, step 8 of 12\n",
      "      2.16 seconds: Optimizing GTR model, step 9 of 12\n",
      "      2.33 seconds: Optimizing GTR model, step 10 of 12\n",
      "      2.51 seconds: Optimizing GTR model, step 11 of 12\n",
      "      2.71 seconds: Optimizing GTR model, step 12 of 12\n",
      "GTR Frequencies: 0.2741 0.2499 0.2693 0.2068\n",
      "GTR rates(ac ag at cg ct gt) 1.4092 2.3564 2.2858 1.6200 5.4781 1.0000\n",
      "      2.90 seconds: ML Lengths 1 of 1702 splits\n",
      "      3.02 seconds: ML Lengths 501 of 1702 splits\n",
      "      3.13 seconds: ML Lengths 1001 of 1702 splits\n",
      "      3.24 seconds: ML Lengths 1501 of 1702 splits\n",
      "      3.35 seconds: Site likelihoods with rate category 3 of 20\n",
      "      3.45 seconds: Site likelihoods with rate category 8 of 20\n",
      "      3.55 seconds: Site likelihoods with rate category 13 of 20\n",
      "      3.65 seconds: Site likelihoods with rate category 18 of 20\n",
      "Switched to using 20 rate categories (CAT approximation)\n",
      "Rate categories were divided by 0.985 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "      3.78 seconds: ML Lengths 201 of 1702 splits\n",
      "      3.88 seconds: ML Lengths 601 of 1702 splits\n",
      "      4.00 seconds: ML Lengths 1101 of 1702 splits\n",
      "      4.11 seconds: ML Lengths 1501 of 1702 splits\n",
      "2 rounds ML lengths: LogLk = -218158.641 Max-change 0.8250 Time 4.17\n",
      "      4.21 seconds: ML Lengths 201 of 1702 splits\n",
      "      4.33 seconds: ML Lengths 701 of 1702 splits\n",
      "      4.45 seconds: ML Lengths 1201 of 1702 splits\n",
      "      4.57 seconds: ML Lengths 1701 of 1702 splits\n",
      "3 rounds ML lengths: LogLk = -218141.351 Max-change 0.1159 Time 4.58\n",
      "      4.67 seconds: ML Lengths 401 of 1702 splits\n",
      "      4.79 seconds: ML Lengths 901 of 1702 splits\n",
      "      4.91 seconds: ML Lengths 1401 of 1702 splits\n",
      "4 rounds ML lengths: LogLk = -218139.979 Max-change 0.0233 Time 4.99\n",
      "      5.01 seconds: ML Lengths 101 of 1702 splits\n",
      "      5.13 seconds: ML Lengths 601 of 1702 splits\n",
      "      5.25 seconds: ML Lengths 1101 of 1702 splits\n",
      "      5.37 seconds: ML Lengths 1601 of 1702 splits\n",
      "5 rounds ML lengths: LogLk = -218139.817 Max-change 0.0052 Time 5.40\n",
      "      5.47 seconds: ML Lengths 301 of 1702 splits\n",
      "      5.58 seconds: ML Lengths 801 of 1702 splits\n",
      "      5.70 seconds: ML Lengths 1301 of 1702 splits\n",
      "6 rounds ML lengths: LogLk = -218139.776 Max-change 0.0020 Time 5.81\n",
      "      5.80 seconds: ML Lengths 1 of 1702 splits\n",
      "      5.92 seconds: ML Lengths 501 of 1702 splits\n",
      "      6.04 seconds: ML Lengths 1001 of 1702 splits\n",
      "      6.15 seconds: ML Lengths 1501 of 1702 splits\n",
      "7 rounds ML lengths: LogLk = -218139.760 Max-change 0.0014 Time 6.21\n",
      "      6.26 seconds: ML Lengths 201 of 1702 splits\n",
      "      6.37 seconds: ML Lengths 701 of 1702 splits\n",
      "      6.49 seconds: ML Lengths 1201 of 1702 splits\n",
      "      6.61 seconds: ML Lengths 1701 of 1702 splits\n",
      "8 rounds ML lengths: LogLk = -218139.751 Max-change 0.0012 Time 6.62\n",
      "      6.71 seconds: ML Lengths 401 of 1702 splits\n",
      "      6.82 seconds: ML Lengths 901 of 1702 splits\n",
      "      6.94 seconds: ML Lengths 1401 of 1702 splits\n",
      "9 rounds ML lengths: LogLk = -218139.746 Max-change 0.0007 (converged) Time 7.02\n",
      "      7.06 seconds: Site likelihoods with rate category 2 of 20\n",
      "      7.17 seconds: Site likelihoods with rate category 7 of 20\n",
      "      7.28 seconds: Site likelihoods with rate category 12 of 20\n",
      "      7.39 seconds: Site likelihoods with rate category 17 of 20\n",
      "Gamma(20) LogLk = -219045.505 alpha = 0.616 rescaling lengths by 1.180\n",
      "Total time: 7.49 seconds Unique: 1704/2769 Bad splits: 0/0\n"
     ]
    }
   ],
   "source": [
    "for gene in genes:\n",
    "    print('>>>>', gene)\n",
    "    indir = os.path.join(workdir, gene)\n",
    "    \n",
    "    # use the original alignment and raxml-ng tree to re-estimate FastTree-2 numeric parameters\n",
    "    rooted_tree_path = os.path.join(indir, 'est.raxml.bestTree.rooted')\n",
    "    alignment_path = os.path.join(indir, 'est.aln.nuc.fasta.gz')\n",
    "    ft_log_path = os.path.join(indir, 'est.fasttree.log')\n",
    "    ft_tree_path = os.path.join(indir, 'est.fasttree.tre')\n",
    "    \n",
    "    # enable multi-thread FastTree-2 (using binary FastTreeMP)\n",
    "    os.system(f\"export OMP_NUM_THREADS={num_threads}\")\n",
    "    \n",
    "    # nucleotide alignment\n",
    "    cmd = ['FastTreeMP', '-nosupport', '-gtr', '-gamma', '-nt']\n",
    "    cmd.extend([\n",
    "        '-mllen', '-nome',\n",
    "        '-log', ft_log_path, '-intree', rooted_tree_path,\n",
    "    ])\n",
    "    \n",
    "    if alignment_path.split('.')[-1] == 'gz':\n",
    "        cmd = ['gzip', '-d', alignment_path, '-c', '|'] + cmd\n",
    "    else:\n",
    "        cmd = ['cat', alignment_path, '|'] + cmd\n",
    "    out_fptr = open(ft_tree_path, 'w')\n",
    "    p = Popen(' '.join(cmd), stdout=out_fptr, stderr=PIPE, shell=True, text=True)#, stdin=in_fptr)\n",
    "    for line in p.stderr:\n",
    "        print(line.replace('\\n', ''))\n",
    "    out_fptr.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3133cfab",
   "metadata": {},
   "source": [
    "# Final step: create the reference package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "83a3649b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 03/18/2025 16:19:46\n",
      "New DB name:   /home/chengzes/Desktop/Research/softwares/TIPP3/refpkg_scripts/custom_tipp_refpkg/markers-v4/blast/alignment.fasta.db\n",
      "New DB title:  custom_tipp_refpkg/markers-v4/blast/alignment.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /home/chengzes/Desktop/Research/softwares/TIPP3/refpkg_scripts/custom_tipp_refpkg/markers-v4/blast/alignment.fasta.db\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 13735 sequences in 0.124663 seconds.\n",
      "\n",
      "\n",
      "total unique taxids: 10960\n",
      "Aggregating all sequences to create a BLAST database...\n",
      "Running makeblastdb...\n",
      "done creating refpkg: RpsM_COG0099\n",
      "done creating refpkg: RplA_COG0081\n",
      "done creating refpkg: RplB_COG0090\n",
      "done creating refpkg: RplD_COG0088\n",
      "done creating refpkg: RpsC_COG0092\n",
      "done computing num sequences: RplB_COG0090\n",
      "done computing num sequences: RplA_COG0081\n",
      "done computing num sequences: RplD_COG0088\n",
      "done computing num sequences: RpsM_COG0099\n",
      "done computing num sequences: RpsC_COG0092\n"
     ]
    }
   ],
   "source": [
    "# run the existing script create_tipp_refpkg.py\n",
    "db_path = os.path.join(workdir, 'ncbi_taxonomy.db')\n",
    "\n",
    "cmd = f\"python create_tipp_refpkg.py {workdir} {db_path} {refpkg_dir} {num_threads}\"\n",
    "p = Popen(cmd, stdout=PIPE, stderr=STDOUT, text=True, shell=True)\n",
    "for line in p.stdout:\n",
    "    print(line.replace('\\n', ''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f78ff556",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
